{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project LAB2 notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing the fasta sequences downloaded from PDB as a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "from Bio.PDB.PDBList import *\n",
    "import os \n",
    "from Bio.PDB.DSSP import dssp_dict_from_pdb_file\n",
    "from Bio.Blast.Applications import NcbipsiblastCommandline\n",
    "import math\n",
    "from sklearn import svm\n",
    "import statistics\n",
    "import joblib\n",
    "from sklearn.metrics import confusion_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing procedure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once downloaded the fasta sequences in *csv* format from PDB I can conver and filter the sequences using the dataframe data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry ID</th>\n",
       "      <th>Entity ID</th>\n",
       "      <th>Chain ID</th>\n",
       "      <th>Database Name</th>\n",
       "      <th>Accession Code(s)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Chain Length</th>\n",
       "      <th>Entity Macromolecule Type</th>\n",
       "      <th>Molecular Weight (Entity)</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5D13</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P31016</td>\n",
       "      <td>GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...</td>\n",
       "      <td>119</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>12.738</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>E, F, G, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XKKETEV</td>\n",
       "      <td>7</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>0.939</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5COQ</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P9WGR1</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...</td>\n",
       "      <td>289</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>30.698</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5CP5</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>O60885</td>\n",
       "      <td>MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...</td>\n",
       "      <td>141</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>16.767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5D15</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>V7LAR8</td>\n",
       "      <td>FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...</td>\n",
       "      <td>177</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>19.587</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28060</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P35444</td>\n",
       "      <td>PSPCHEKADCILERDGSRS</td>\n",
       "      <td>19</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>2.12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28061</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>A, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...</td>\n",
       "      <td>233</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>25.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28062</th>\n",
       "      <td>6SET</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28063</th>\n",
       "      <td>6SEW</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28064</th>\n",
       "      <td>6SEV</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D, E, F</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P80725</td>\n",
       "      <td>VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...</td>\n",
       "      <td>150</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>17.395</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28065 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Entry ID Entity ID          Chain ID Database Name Accession Code(s)  \\\n",
       "0         5D13         1        A, B, C, D       UniProt            P31016   \n",
       "1          NaN         2        E, F, G, H           NaN               NaN   \n",
       "2         5COQ         1        A, B, C, D       UniProt            P9WGR1   \n",
       "3         5CP5         1                 A       UniProt            O60885   \n",
       "4         5D15         1              A, B       UniProt            V7LAR8   \n",
       "...        ...       ...               ...           ...               ...   \n",
       "28060      NaN         3              C, D       UniProt            P35444   \n",
       "28061      NaN         1              A, H           NaN               NaN   \n",
       "28062     6SET         1                 A       UniProt            P00698   \n",
       "28063     6SEW         1                 A       UniProt            P00698   \n",
       "28064     6SEV         1  A, B, C, D, E, F       UniProt            P80725   \n",
       "\n",
       "                                                Sequence Chain Length  \\\n",
       "0      GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...          119   \n",
       "1                                                XKKETEV            7   \n",
       "2      MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...          289   \n",
       "3      MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...          141   \n",
       "4      FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...          177   \n",
       "...                                                  ...          ...   \n",
       "28060                                PSPCHEKADCILERDGSRS           19   \n",
       "28061  QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...          233   \n",
       "28062  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28063  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28064  VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...          150   \n",
       "\n",
       "      Entity Macromolecule Type Molecular Weight (Entity)  Unnamed: 9  \n",
       "0                polypeptide(L)                    12.738         NaN  \n",
       "1                polypeptide(L)                     0.939         NaN  \n",
       "2                polypeptide(L)                    30.698         NaN  \n",
       "3                polypeptide(L)                    16.767         NaN  \n",
       "4                polypeptide(L)                    19.587         NaN  \n",
       "...                         ...                       ...         ...  \n",
       "28060            polypeptide(L)                      2.12         NaN  \n",
       "28061            polypeptide(L)                     25.39         NaN  \n",
       "28062            polypeptide(L)                    14.331         NaN  \n",
       "28063            polypeptide(L)                    14.331         NaN  \n",
       "28064            polypeptide(L)                    17.395         NaN  \n",
       "\n",
       "[28065 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df=pd.read_csv(\"fasta_files.csv\")\n",
    "fasta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I want to replace those sequences that have a NaN with the ID of the sequence that is above it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry ID</th>\n",
       "      <th>Entity ID</th>\n",
       "      <th>Chain ID</th>\n",
       "      <th>Database Name</th>\n",
       "      <th>Accession Code(s)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Chain Length</th>\n",
       "      <th>Entity Macromolecule Type</th>\n",
       "      <th>Molecular Weight (Entity)</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5D13</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P31016</td>\n",
       "      <td>GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...</td>\n",
       "      <td>119</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>12.738</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5D13</td>\n",
       "      <td>2</td>\n",
       "      <td>E, F, G, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XKKETEV</td>\n",
       "      <td>7</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>0.939</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5COQ</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P9WGR1</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...</td>\n",
       "      <td>289</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>30.698</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5CP5</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>O60885</td>\n",
       "      <td>MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...</td>\n",
       "      <td>141</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>16.767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5D15</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>V7LAR8</td>\n",
       "      <td>FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...</td>\n",
       "      <td>177</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>19.587</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28060</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>3</td>\n",
       "      <td>C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P35444</td>\n",
       "      <td>PSPCHEKADCILERDGSRS</td>\n",
       "      <td>19</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>2.12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28061</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>1</td>\n",
       "      <td>A, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...</td>\n",
       "      <td>233</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>25.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28062</th>\n",
       "      <td>6SET</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28063</th>\n",
       "      <td>6SEW</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28064</th>\n",
       "      <td>6SEV</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D, E, F</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P80725</td>\n",
       "      <td>VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...</td>\n",
       "      <td>150</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>17.395</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28065 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Entry ID Entity ID          Chain ID Database Name Accession Code(s)  \\\n",
       "0         5D13         1        A, B, C, D       UniProt            P31016   \n",
       "1         5D13         2        E, F, G, H           NaN               NaN   \n",
       "2         5COQ         1        A, B, C, D       UniProt            P9WGR1   \n",
       "3         5CP5         1                 A       UniProt            O60885   \n",
       "4         5D15         1              A, B       UniProt            V7LAR8   \n",
       "...        ...       ...               ...           ...               ...   \n",
       "28060     6SF6         3              C, D       UniProt            P35444   \n",
       "28061     6SF6         1              A, H           NaN               NaN   \n",
       "28062     6SET         1                 A       UniProt            P00698   \n",
       "28063     6SEW         1                 A       UniProt            P00698   \n",
       "28064     6SEV         1  A, B, C, D, E, F       UniProt            P80725   \n",
       "\n",
       "                                                Sequence Chain Length  \\\n",
       "0      GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...          119   \n",
       "1                                                XKKETEV            7   \n",
       "2      MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...          289   \n",
       "3      MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...          141   \n",
       "4      FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...          177   \n",
       "...                                                  ...          ...   \n",
       "28060                                PSPCHEKADCILERDGSRS           19   \n",
       "28061  QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...          233   \n",
       "28062  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28063  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28064  VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...          150   \n",
       "\n",
       "      Entity Macromolecule Type Molecular Weight (Entity)  Unnamed: 9  \n",
       "0                polypeptide(L)                    12.738         NaN  \n",
       "1                polypeptide(L)                     0.939         NaN  \n",
       "2                polypeptide(L)                    30.698         NaN  \n",
       "3                polypeptide(L)                    16.767         NaN  \n",
       "4                polypeptide(L)                    19.587         NaN  \n",
       "...                         ...                       ...         ...  \n",
       "28060            polypeptide(L)                      2.12         NaN  \n",
       "28061            polypeptide(L)                     25.39         NaN  \n",
       "28062            polypeptide(L)                    14.331         NaN  \n",
       "28063            polypeptide(L)                    14.331         NaN  \n",
       "28064            polypeptide(L)                    17.395         NaN  \n",
       "\n",
       "[28065 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for index in fasta_df.index:\n",
    "    #print(index)\n",
    "    if pd.isnull(fasta_df.loc[index,\"Entry ID\"]):\n",
    "        #print(fasta_df.loc[index,\"Entry ID\"])\n",
    "        fasta_df.loc[index,\"Entry ID\"]=fasta_df.loc[index-1,\"Entry ID\"]\n",
    "#fasta_df=fasta_df[fasta_df[\"Entry ID\"].notna()]\n",
    "fasta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I want to delete those sequences that have less than 50 and more than 300 aminoacids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry ID</th>\n",
       "      <th>Entity ID</th>\n",
       "      <th>Chain ID</th>\n",
       "      <th>Database Name</th>\n",
       "      <th>Accession Code(s)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Chain Length</th>\n",
       "      <th>Entity Macromolecule Type</th>\n",
       "      <th>Molecular Weight (Entity)</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5D13</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P31016</td>\n",
       "      <td>GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...</td>\n",
       "      <td>119</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>12.738</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5COQ</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P9WGR1</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...</td>\n",
       "      <td>289</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>30.698</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5CP5</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>O60885</td>\n",
       "      <td>MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...</td>\n",
       "      <td>141</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>16.767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5D15</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>V7LAR8</td>\n",
       "      <td>FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...</td>\n",
       "      <td>177</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>19.587</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5COS</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>A0A084CH09</td>\n",
       "      <td>GAMGMNGQGATSIPGEVAEQAMHWHLELQEPAVSAATLAACMSWRQ...</td>\n",
       "      <td>86</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>9.515</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28059</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>2</td>\n",
       "      <td>B, L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DVLMTQIPLSLPVSLGDQASISCRSSQNIVHSNGNTYLEWYLQKPG...</td>\n",
       "      <td>219</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>24.236</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28061</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>1</td>\n",
       "      <td>A, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...</td>\n",
       "      <td>233</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>25.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28062</th>\n",
       "      <td>6SET</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28063</th>\n",
       "      <td>6SEW</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28064</th>\n",
       "      <td>6SEV</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D, E, F</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P80725</td>\n",
       "      <td>VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...</td>\n",
       "      <td>150</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>17.395</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24066 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Entry ID Entity ID          Chain ID Database Name Accession Code(s)  \\\n",
       "0         5D13         1        A, B, C, D       UniProt            P31016   \n",
       "2         5COQ         1        A, B, C, D       UniProt            P9WGR1   \n",
       "3         5CP5         1                 A       UniProt            O60885   \n",
       "4         5D15         1              A, B       UniProt            V7LAR8   \n",
       "5         5COS         1        A, B, C, D       UniProt        A0A084CH09   \n",
       "...        ...       ...               ...           ...               ...   \n",
       "28059     6SF6         2              B, L           NaN               NaN   \n",
       "28061     6SF6         1              A, H           NaN               NaN   \n",
       "28062     6SET         1                 A       UniProt            P00698   \n",
       "28063     6SEW         1                 A       UniProt            P00698   \n",
       "28064     6SEV         1  A, B, C, D, E, F       UniProt            P80725   \n",
       "\n",
       "                                                Sequence Chain Length  \\\n",
       "0      GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...          119   \n",
       "2      MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...          289   \n",
       "3      MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...          141   \n",
       "4      FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...          177   \n",
       "5      GAMGMNGQGATSIPGEVAEQAMHWHLELQEPAVSAATLAACMSWRQ...           86   \n",
       "...                                                  ...          ...   \n",
       "28059  DVLMTQIPLSLPVSLGDQASISCRSSQNIVHSNGNTYLEWYLQKPG...          219   \n",
       "28061  QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...          233   \n",
       "28062  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28063  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "28064  VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...          150   \n",
       "\n",
       "      Entity Macromolecule Type Molecular Weight (Entity)  Unnamed: 9  \n",
       "0                polypeptide(L)                    12.738         NaN  \n",
       "2                polypeptide(L)                    30.698         NaN  \n",
       "3                polypeptide(L)                    16.767         NaN  \n",
       "4                polypeptide(L)                    19.587         NaN  \n",
       "5                polypeptide(L)                     9.515         NaN  \n",
       "...                         ...                       ...         ...  \n",
       "28059            polypeptide(L)                    24.236         NaN  \n",
       "28061            polypeptide(L)                     25.39         NaN  \n",
       "28062            polypeptide(L)                    14.331         NaN  \n",
       "28063            polypeptide(L)                    14.331         NaN  \n",
       "28064            polypeptide(L)                    17.395         NaN  \n",
       "\n",
       "[24066 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#REMOVING FIRST THE SEQUENCES THAT ARE LESS THAN 50 AA LONG\n",
    "#pd.to_numeric(fasta_df[\"Chain Length\"],errors=\"coerce\")\n",
    "#fasta_df_length=fasta_df[pd.to_numeric(fasta_df[\"Chain Length\"],errors=\"coerce\")>50]\n",
    "fasta_df=fasta_df[pd.to_numeric(fasta_df[\"Chain Length\"],errors=\"coerce\")>50]\n",
    "#REMOVING THEN THE SEQUENCES THAT ARE MORE THAN 300 AA LONG\n",
    "fasta_df=fasta_df[pd.to_numeric(fasta_df[\"Chain Length\"],errors=\"coerce\")<300]\n",
    "fasta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I want now to remove all those sequences that have the X in their AA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry ID</th>\n",
       "      <th>Entity ID</th>\n",
       "      <th>Chain ID</th>\n",
       "      <th>Database Name</th>\n",
       "      <th>Accession Code(s)</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Chain Length</th>\n",
       "      <th>Entity Macromolecule Type</th>\n",
       "      <th>Molecular Weight (Entity)</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5D13</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P31016</td>\n",
       "      <td>GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...</td>\n",
       "      <td>119</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>12.738</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5COQ</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P9WGR1</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...</td>\n",
       "      <td>289</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>30.698</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5CP5</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>O60885</td>\n",
       "      <td>MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...</td>\n",
       "      <td>141</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>16.767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5D15</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>V7LAR8</td>\n",
       "      <td>FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...</td>\n",
       "      <td>177</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>19.587</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5COS</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>A0A084CH09</td>\n",
       "      <td>GAMGMNGQGATSIPGEVAEQAMHWHLELQEPAVSAATLAACMSWRQ...</td>\n",
       "      <td>86</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>9.515</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23868</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>2</td>\n",
       "      <td>B, L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DVLMTQIPLSLPVSLGDQASISCRSSQNIVHSNGNTYLEWYLQKPG...</td>\n",
       "      <td>219</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>24.236</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23869</th>\n",
       "      <td>6SF6</td>\n",
       "      <td>1</td>\n",
       "      <td>A, H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...</td>\n",
       "      <td>233</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>25.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23870</th>\n",
       "      <td>6SET</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23871</th>\n",
       "      <td>6SEW</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P00698</td>\n",
       "      <td>KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...</td>\n",
       "      <td>129</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>14.331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23872</th>\n",
       "      <td>6SEV</td>\n",
       "      <td>1</td>\n",
       "      <td>A, B, C, D, E, F</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>P80725</td>\n",
       "      <td>VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...</td>\n",
       "      <td>150</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>17.395</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23873 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Entry ID Entity ID          Chain ID Database Name Accession Code(s)  \\\n",
       "0         5D13         1        A, B, C, D       UniProt            P31016   \n",
       "1         5COQ         1        A, B, C, D       UniProt            P9WGR1   \n",
       "2         5CP5         1                 A       UniProt            O60885   \n",
       "3         5D15         1              A, B       UniProt            V7LAR8   \n",
       "4         5COS         1        A, B, C, D       UniProt        A0A084CH09   \n",
       "...        ...       ...               ...           ...               ...   \n",
       "23868     6SF6         2              B, L           NaN               NaN   \n",
       "23869     6SF6         1              A, H           NaN               NaN   \n",
       "23870     6SET         1                 A       UniProt            P00698   \n",
       "23871     6SEW         1                 A       UniProt            P00698   \n",
       "23872     6SEV         1  A, B, C, D, E, F       UniProt            P80725   \n",
       "\n",
       "                                                Sequence Chain Length  \\\n",
       "0      GSPEFLGEEDIPREPRRIVIHRGSTGLGFNIVGGEDGEGIFISFIL...          119   \n",
       "1      MGSSHHHHHHSSGLVPRGSHMTGLLDGKRILVSGIITDSSIAFHIA...          289   \n",
       "2      MKKGHHHHHHLVPRGSNPPPPETSNPNKPKRQTNQLQYLLRVVLKT...          141   \n",
       "3      FQGAMGSRVVILFTDIEESTALNERIGDRAWVKLISSHDKLVSDLV...          177   \n",
       "4      GAMGMNGQGATSIPGEVAEQAMHWHLELQEPAVSAATLAACMSWRQ...           86   \n",
       "...                                                  ...          ...   \n",
       "23868  DVLMTQIPLSLPVSLGDQASISCRSSQNIVHSNGNTYLEWYLQKPG...          219   \n",
       "23869  QVQLQQSGAELAKPGASVNLSCKASGYTFTNYWVHWVKQRPGQGLE...          233   \n",
       "23870  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "23871  KVFGRCELAAAMKRHGLDNYRGYSLGNWVCAAKFESNFNTQATNRN...          129   \n",
       "23872  VDTKEFLNHQVANLNVFTVKIHQIHWYMRGHNFFTLHEKMDDLYSE...          150   \n",
       "\n",
       "      Entity Macromolecule Type Molecular Weight (Entity)  Unnamed: 9  \n",
       "0                polypeptide(L)                    12.738         NaN  \n",
       "1                polypeptide(L)                    30.698         NaN  \n",
       "2                polypeptide(L)                    16.767         NaN  \n",
       "3                polypeptide(L)                    19.587         NaN  \n",
       "4                polypeptide(L)                     9.515         NaN  \n",
       "...                         ...                       ...         ...  \n",
       "23868            polypeptide(L)                    24.236         NaN  \n",
       "23869            polypeptide(L)                     25.39         NaN  \n",
       "23870            polypeptide(L)                    14.331         NaN  \n",
       "23871            polypeptide(L)                    14.331         NaN  \n",
       "23872            polypeptide(L)                    17.395         NaN  \n",
       "\n",
       "[23873 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df=fasta_df[fasta_df[\"Sequence\"].str.contains(\"X\")==False]\n",
    "fasta_df.reset_index(drop=True, inplace=True)\n",
    "fasta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clusterization procedure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to perform the clusterization procedure, but I have to convert the DataFrame in a multifasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "touch: missing file operand\n",
      "Try 'touch --help' for more information.\n"
     ]
    }
   ],
   "source": [
    "!touch > fasta.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file=open(\"fasta.fasta\",\"a\")\n",
    "for index in fasta_df.index:\n",
    "    PDB_ID=\">\"+fasta_df.loc[index,\"Entry ID\"]+\"_\"+fasta_df.loc[index,\"Chain ID\"].split(\",\")[0]\n",
    "    Sequence=fasta_df.loc[index,\"Sequence\"]\n",
    "    print(PDB_ID,file=fasta_file)\n",
    "    #print(\"\\n\",file=fasta_file)\n",
    "    print(Sequence,file=fasta_file)    \n",
    "#fasta_file.write(fasta_df.to_string(fasta_df[\"Sequence\"]))\n",
    "fasta_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  23873   23873  191260\n"
     ]
    }
   ],
   "source": [
    "!grep \">\" fasta.fasta | wc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to run the mmseqs clusterization package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster fasta.fasta Clustering_res tmp --min-seq-id 0.3 -c 0.5 --cluster-mode 0 \n",
      "\n",
      "MMseqs Version:                     \t12.113e3\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.5\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t4\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Realign hits                        \tfalse\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb fasta.fasta tmp/2056378403674707787/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[23837] 0s 28ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 30ms\n",
      "\u001b[33mTmp tmp/2056378403674707787/clu_tmp folder does not exist or is not a directory.\n",
      "\u001b[39mCreate dir tmp/2056378403674707787/clu_tmp\n",
      "cluster tmp/2056378403674707787/input tmp/2056378403674707787/clu tmp/2056378403674707787/clu_tmp --max-seqs 20 -c 0.5 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.3 --cluster-mode 0 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster iterations to 3\n",
      "linclust tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --realign 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.5 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.5 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 23873 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 23.87K 0s 161ms    \n",
      "Sort kmer 0h 0m 0s 54ms\n",
      "Sort by rep. sequence 0h 0m 0s 20ms\n",
      "Time for fill: 0h 0m 0s 10ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 379ms\n",
      "rescorediagonal tmp/2056378403674707787/input tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.5 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 23.87K 0s 19ms     \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 20ms\n",
      "Time for processing: 0h 0m 0s 62ms\n",
      "clust tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_rescore1 tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 23.87K 0s 18ms     \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 40644 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 23.87K 0s 20ms     \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 23.87K 0s 2ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 85ms\n",
      "Total time: 0h 0m 0s 117ms\n",
      "\n",
      "Size of the sequence database: 23873\n",
      "Size of the alignment database: 23873\n",
      "Number of clusters: 6183\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 130ms\n",
      "createsubdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/order_redundancy tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "createsubdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/order_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 60ms\n",
      "filterdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter1 tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter2 --filter-file tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/order_redundancy --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 6.18K 0s 9ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "rescorediagonal tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter2 tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.5 -a 0 --cov-mode 0 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 6.18K 0s 5ms      \n",
      "Time for merging to pref_rescore2: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 20ms\n",
      "align tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_rescore2 tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --realign 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 6183 type: Aminoacid\n",
      "Target database size: 6183 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 6.18K 0s 176ms    \n",
      "Time for merging to aln: 0h 0m 0s 4ms\n",
      "\n",
      "6984 alignments calculated.\n",
      "6748 sequence pairs passed the thresholds (0.966208 of overall calculated).\n",
      "1.091380 hits per query sequence.\n",
      "Time for processing: 0h 0m 0s 187ms\n",
      "clust tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/aln tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 6.18K 0s 9ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 565 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 6.18K 0s 15ms     ============>                                                ] 25.69% 1.59K eta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 6.18K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 27ms\n",
      "Total time: 0h 0m 0s 28ms\n",
      "\n",
      "Size of the sequence database: 6183\n",
      "Size of the alignment database: 6183\n",
      "Number of clusters: 5780\n",
      "\n",
      "Writing results 0h 0m 0s 14ms\n",
      "Time for merging to clust: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 53ms\n",
      "mergeclusters tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pre_clust tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/clust --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 6.18K 0s 3ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 5.78K 0s 9ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 23.87K 0s 31ms     \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 0s 99ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/linclust/13895547741012165956/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_redundancy tmp/2056378403674707787/input tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "prefilter tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 5780 type: Aminoacid\n",
      "Estimated memory consumption: 985M\n",
      "Target database size: 5780 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 5.78K 0s 96ms     \n",
      "Index table: Masked residues: 17436\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 5.78K 0s 56ms     \n",
      "Index statistics\n",
      "Entries:          460578\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.007197\n",
      "Top 10 k-mers\n",
      "    SGVRHM\t211\n",
      "    VDGEYF\t39\n",
      "    DLTNFQ\t39\n",
      "    SMGQGR\t29\n",
      "    GTNYSM\t24\n",
      "    YWQTTV\t24\n",
      "    SGPRGS\t21\n",
      "    WGGQVS\t21\n",
      "    PRSMMT\t21\n",
      "    RQPKEW\t16\n",
      "Time for index table init: 0h 0m 0s 845ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 5780\n",
      "Target db start 1 to 5780\n",
      "[=================================================================] 100.00% 5.78K 0s 214ms    \n",
      "\n",
      "2.151443 k-mers per position\n",
      "99 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 2s 444ms\n",
      "align tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step0 tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --realign 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 5780 type: Aminoacid\n",
      "Target database size: 5780 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 5.78K 0s 368ms    \n",
      "Time for merging to aln_step0: 0h 0m 0s 3ms\n",
      "\n",
      "14444 alignments calculated.\n",
      "9140 sequence pairs passed the thresholds (0.632789 of overall calculated).\n",
      "1.581315 hits per query sequence.\n",
      "Time for processing: 0h 0m 0s 384ms\n",
      "clust tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step0 tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 5.78K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 248 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 5.78K 0s 4ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 5.78K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 13ms\n",
      "Total time: 0h 0m 0s 14ms\n",
      "\n",
      "Size of the sequence database: 5780\n",
      "Size of the alignment database: 5780\n",
      "Number of clusters: 4938\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 29ms\n",
      "createsubdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step0 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 3 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4938 type: Aminoacid\n",
      "Estimated memory consumption: 984M\n",
      "Target database size: 4938 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 4.94K 0s 91ms     \n",
      "Index table: Masked residues: 14911\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 4.94K 0s 72ms     \n",
      "Index statistics\n",
      "Entries:          764034\n",
      "DB size:          492 MB\n",
      "Avg k-mer size:   0.011938\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t167\n",
      "    SGVRHM\t163\n",
      "    SGDGNL\t27\n",
      "    DLTNFQ\t27\n",
      "    LGELQS\t27\n",
      "    GVLTLY\t27\n",
      "    GLPGMT\t15\n",
      "    HMSTQQ\t13\n",
      "    PRSMMT\t13\n",
      "    GLPGMV\t12\n",
      "Time for index table init: 0h 0m 0s 840ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4938\n",
      "Target db start 1 to 4938\n",
      "[=================================================================] 100.00% 4.94K 0s 862ms    \n",
      "\n",
      "22.842622 k-mers per position\n",
      "196 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 2s 932ms\n",
      "align tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --realign 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4938 type: Aminoacid\n",
      "Target database size: 4938 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 4.94K 0s 282ms    \n",
      "Time for merging to aln_step1: 0h 0m 0s 3ms\n",
      "\n",
      "11344 alignments calculated.\n",
      "6159 sequence pairs passed the thresholds (0.542930 of overall calculated).\n",
      "1.247266 hits per query sequence.\n",
      "Time for processing: 0h 0m 0s 321ms\n",
      "clust tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 4.94K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 165 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 4.94K 0s 3ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 4.94K 0s 5ms      ====================>                                      ] 40.51% 2.00K eta 0s       \n",
      "\n",
      "Time for read in: 0h 0m 0s 16ms\n",
      "Total time: 0h 0m 0s 22ms\n",
      "\n",
      "Size of the sequence database: 4938\n",
      "Size of the alignment database: 4938\n",
      "Number of clusters: 4401\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 40ms\n",
      "createsubdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "prefilter tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 5 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.5 --cov-mode 0 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4401 type: Aminoacid\n",
      "Estimated memory consumption: 983M\n",
      "Target database size: 4401 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 4.40K 0s 83ms     \n",
      "Index table: Masked residues: 13317\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 4.40K 0s 77ms     \n",
      "Index statistics\n",
      "Entries:          702224\n",
      "DB size:          492 MB\n",
      "Avg k-mer size:   0.010972\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t134\n",
      "    SGVRHM\t131\n",
      "    SGDGNL\t24\n",
      "    SSVLEN\t24\n",
      "    DLTNFQ\t24\n",
      "    LGELQS\t24\n",
      "    GVLTLY\t24\n",
      "    GLPGMT\t13\n",
      "    YDPTLY\t12\n",
      "    TGQMGS\t10\n",
      "Time for index table init: 0h 0m 0s 856ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4401\n",
      "Target db start 1 to 4401\n",
      "[=================================================================] 100.00% 4.40K 3s 236ms    \n",
      "\n",
      "171.110488 k-mers per position\n",
      "463 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "4 sequences passed prefiltering per query sequence\n",
      "3 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 5s 331ms\n",
      "align tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.5 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --realign 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4401 type: Aminoacid\n",
      "Target database size: 4401 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 4.40K 0s 279ms    \n",
      "Time for merging to aln_step2: 0h 0m 0s 7ms\n",
      "\n",
      "17589 alignments calculated.\n",
      "5043 sequence pairs passed the thresholds (0.286713 of overall calculated).\n",
      "1.145876 hits per query sequence.\n",
      "Time for processing: 0h 0m 0s 302ms\n",
      "clust tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step2 tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 4.40K 0s 2ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 116 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 4.40K 0s 3ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 4.40K 0s 0ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 7ms\n",
      "Total time: 0h 0m 0s 11ms\n",
      "\n",
      "Size of the sequence database: 4401\n",
      "Size of the alignment database: 4401\n",
      "Number of clusters: 4086\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "mergeclusters tmp/2056378403674707787/input tmp/2056378403674707787/clu tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_redundancy tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step0 tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step1 tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step2 --threads 4 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 5.78K 0s 10ms     \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 4.94K 0s 45ms     \n",
      "Clustering step 3\n",
      "[=================================================================] 100.00% 4.40K 0s 47ms     \n",
      "Clustering step 4\n",
      "[=================================================================] 100.00% 4.09K 0s 50ms     .97% 3.72K eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 23.87K 0s 60ms     \n",
      "Time for merging to clu: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 79ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "rmdb tmp/2056378403674707787/clu_tmp/3380204455719165468/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv tmp/2056378403674707787/input tmp/2056378403674707787/input tmp/2056378403674707787/clu tmp/2056378403674707787/cluster.tsv --threads 4 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 11ms\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "result2repseq tmp/2056378403674707787/input tmp/2056378403674707787/clu tmp/2056378403674707787/clu_rep --threads 4 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 4.09K 0s 3ms      \n",
      "Time for merging to clu_rep: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "result2flat tmp/2056378403674707787/input tmp/2056378403674707787/input tmp/2056378403674707787/clu_rep tmp/2056378403674707787/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 51ms\n",
      "createseqfiledb tmp/2056378403674707787/input tmp/2056378403674707787/clu tmp/2056378403674707787/clu_seqs --threads 4 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 4.09K 0s 8ms      \n",
      "Time for merging to clu_seqs: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 65ms\n",
      "result2flat tmp/2056378403674707787/input tmp/2056378403674707787/input tmp/2056378403674707787/clu_seqs tmp/2056378403674707787/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 83ms\n",
      "rmdb tmp/2056378403674707787/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/2056378403674707787/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/2056378403674707787/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster fasta.fasta Clustering_res tmp --min-seq-id 0.3 -c 0.5 --cluster-mode 0 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4086\n"
     ]
    }
   ],
   "source": [
    "!grep \">\" Clustering_res_rep_seq.fasta | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file in which we are interested is the \"Clustering_res_rep_seq.fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing redundancy between sets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have now to open the \"Tot_sequence_JPred4.fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now that I have my Blind set of representative protein sequences and the JPred4 fasta I can remove the redundacy between these two sets by blasting them one against the other. And I want to retain only those proteins that have low sequence identity.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to make the Tot_sequence_JPred4.fasta as the database for blasting against it the Blindset (\"Clustering_res_rep_seq.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 09/23/2020 01:03:10\n",
      "New DB name:   /mnt/EA7C9C107C9BD5A3/User/BIOINFORMATICA.BOLOGNA/LAB_2/Project/Project2/Tot_sequence_JPred4.fasta\n",
      "New DB title:  Tot_sequence_JPred4.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /mnt/EA7C9C107C9BD5A3/User/BIOINFORMATICA.BOLOGNA/LAB_2/Project/Project2/Tot_sequence_JPred4.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1348 sequences in 0.142454 seconds.\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in Tot_sequence_JPred4.fasta -dbtype prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "!blastp -query Clustering_res_rep_seq.fasta -db Tot_sequence_JPred4.fasta -evalue 0.01 -out hits_blast.tab -outfmt 6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The generated file is the \"hits_blast.tab\" "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to remove the redundancy between the two datasets by removing all the sequences that have more than 30% of sequence identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "555\n"
     ]
    }
   ],
   "source": [
    "hits=open(\"hits_blast.tab\")\n",
    "list_ID=[]\n",
    "for line in hits:\n",
    "    list_line=[]\n",
    "    list_line=line.split(\"\\t\")\n",
    "    if float(list_line[2])>30.00:\n",
    "        if list_line[0] not in list_ID:\n",
    "             list_ID.append(list_line[0])\n",
    "print(len(list_ID))\n",
    "#print(list_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the list_ID there are all the fasta ID that we want to remove from the 4086 proteinf in the Clustering_res_rep_seq.fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3532\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4XXP_A</td>\n",
       "      <td>MAHHHHHHASPAPPPAAPSGQSGTYLFQDEFDGPAGSAPDASKWAV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4XXF_A</td>\n",
       "      <td>GGLAALTKPPTFATVEAERAWLKERLVAAIRIFANEGFDHTVAGHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4XXI_A</td>\n",
       "      <td>MLAVATITQAEQQDRFLGRGELDELASYFASGAKRLEIAQLLTENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4Y9I_A</td>\n",
       "      <td>GVHVLDRPIVLFTTTGAKSGKKRYVPLMRVEENGKYAMVASKGGDP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4XXB_A</td>\n",
       "      <td>MAQDQGEKENPMRELRIRKLCLNICVGESGDRLTRAAKVLEQLTGQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3527</th>\n",
       "      <td>6M89_A</td>\n",
       "      <td>GSFTSHQVAGHMYGKDKVGILQHPDGTVLKQLQPPPRGPRELEFYN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3528</th>\n",
       "      <td>6LW3_A</td>\n",
       "      <td>GPLGSGRPKKMTLILGIDPGSRITGFGVVRETARGCEYVASGCIRT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3529</th>\n",
       "      <td>6M96_B</td>\n",
       "      <td>MNLSLILELVRQEIKNRYADTVLGIWWAFLWPILLVLIYTLIFSHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3530</th>\n",
       "      <td>6M9K_D</td>\n",
       "      <td>ITPVNDETMQEINTLLIALDKTWDDDLLPLCSQIFRRDIRASSELT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3531</th>\n",
       "      <td>6M9K_A</td>\n",
       "      <td>MTPDIILQRTGIDVRAVEQGDDAWHKLRLGVITASEVHNVIAKPRS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3532 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID                                           Sequence\n",
       "0     4XXP_A  MAHHHHHHASPAPPPAAPSGQSGTYLFQDEFDGPAGSAPDASKWAV...\n",
       "1     4XXF_A  GGLAALTKPPTFATVEAERAWLKERLVAAIRIFANEGFDHTVAGHL...\n",
       "2     4XXI_A  MLAVATITQAEQQDRFLGRGELDELASYFASGAKRLEIAQLLTENS...\n",
       "3     4Y9I_A  GVHVLDRPIVLFTTTGAKSGKKRYVPLMRVEENGKYAMVASKGGDP...\n",
       "4     4XXB_A  MAQDQGEKENPMRELRIRKLCLNICVGESGDRLTRAAKVLEQLTGQ...\n",
       "...      ...                                                ...\n",
       "3527  6M89_A  GSFTSHQVAGHMYGKDKVGILQHPDGTVLKQLQPPPRGPRELEFYN...\n",
       "3528  6LW3_A  GPLGSGRPKKMTLILGIDPGSRITGFGVVRETARGCEYVASGCIRT...\n",
       "3529  6M96_B  MNLSLILELVRQEIKNRYADTVLGIWWAFLWPILLVLIYTLIFSHL...\n",
       "3530  6M9K_D  ITPVNDETMQEINTLLIALDKTWDDDLLPLCSQIFRRDIRASSELT...\n",
       "3531  6M9K_A  MTPDIILQRTGIDVRAVEQGDDAWHKLRLGVITASEVHNVIAKPRS...\n",
       "\n",
       "[3532 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep=open(\"Clustering_res_rep_seq.fasta\",\"r\")\n",
    "rep_dict={}\n",
    "rep_ID=[]\n",
    "for line in rep:\n",
    "    if line.startswith(\">\"):\n",
    "        #print(line[1:7])\n",
    "        line=line.rstrip()[1:7]\n",
    "        #print(line)\n",
    "        if line not in list_ID:\n",
    "            keys=rep_dict.keys()\n",
    "            if line not in rep_ID:\n",
    "                rep_ID.append(line)\n",
    "                if line not in keys:\n",
    "                    rep_dict[line]=rep.readline().rstrip()    \n",
    "rep_df=pd.DataFrame.from_dict(rep_dict,orient=\"index\") #IMPORTANT SYNTAX!!\n",
    "rep_df.reset_index(inplace=True) #IMPORTANT SYNTAX!!!\n",
    "rep_df.columns=[\"ID\",\"Sequence\"] #IMPORTANT SYNTAX!!!\n",
    "print(len(rep_ID))\n",
    "rep_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the rep_ID there are all the ID of the representative set (our blind set) from which we are now going to randomly extract 150 sequence ID that we want to download to PDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Donwload PDB files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will now randomly extract only 150 PDB ID from the rep_ID list that I will then download from PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "150\n",
      "['4XXT_A', '5XDH_A', '5DJT_B', '4ZV9_A', '5ZZ6_A', '5Z48_A', '6NMW_A', '5EZU_A', '4YSL_A', '5IYS_A', '5HX4_A', '5GKB_A', '5VKD_A', '6KSY_A', '6EWH_A', '6IN9_C', '5YGE_A', '4XU4_A', '6K2H_A', '6ELU_A', '5N2I_A', '6A45_A', '5GKK_A', '6GJE_B', '6HRR_A', '4YH1_A', '6JCC_A', '4YHV_A', '5VHT_A', '6QJA_A', '4YMR_A', '6SPZ_A', '6O2V_A', '5VJT_A', '6J2Z_A', '6SSH_A', '6ESK_A', '5VYM_A', '6BDC_A', '5OHF_A', '4XXF_A', '6DN5_A', '5HI3_A', '5ME7_A', '5FCU_G', '5FSB_A', '6I6J_A', '6GOS_A', '5DY9_A', '5X6L_A', '6PUP_A', '4YUE_C', '6W2L_B', '5E5M_A', '5FV5_A', '6IN8_A', '6OM5_A', '5Z4Y_A', '5OIY_A', '5FEB_A', '5NFD_A', '6DTC_A', '6OF0_A', '5JO8_A', '6FDD_A', '5IQJ_A', '4XZX_A', '6NHL_A', '4YK1_A', '5YHH_A', '4YAP_A', '6T84_A', '6HAT_A', '6QO1_A', '6MD3_A', '6TUM_A', '6YMU_B', '6W04_A', '5H5E_A', '6KN5_A', '5LXO_A', '6FLM_A', '5X7L_A', '4YBH_B', '6AHT_A', '5FAI_A', '5F7T_E', '6I8B_B', '5B4B_A', '6L4P_B', '5Y21_A', '6H4K_A', '6F3D_A', '5DQS_A', '5H9K_A', '7C2G_G', '4ZC4_A', '5JR5_A', '6FPK_A', '6BP6_A', '6EEL_A', '6W3F_A', '5CAJ_A', '5DIF_B', '6M96_B', '5TTG_A', '5V23_A', '5T50_A', '4ZSI_A', '5PWQ_A', '5HTL_A', '5T87_E', '5YEI_B', '5UNI_A', '6O5K_A', '5UDQ_A', '4XXB_A', '6BK4_A', '5U9C_A', '4XJC_A', '5TF4_A', '5DHE_A', '5WHF_A', '6B6I_A', '6S8X_A', '6GIE_A', '5U4U_A', '6KZR_A', '5NE4_3', '6LJ9_A', '6FXF_A', '6SJQ_A', '5TKA_A', '4ZXG_A', '5JRV_A', '6MNU_A', '5XHB_A', '5T7H_A', '5TED_A', '5CYV_A', '5VOG_A', '6FP5_A', '6LTZ_A', '5H4G_A', '5W95_A', '4ZLX_A', '6PLJ_A', '5MYD_A', '6RFG_A', '5FOH_A']\n"
     ]
    }
   ],
   "source": [
    "#import random\n",
    "def randomize(rep_ID):\n",
    "    randomlist=[]\n",
    "    random.seed(22092020) # RANDOM SEED!!!!!!!!!!!!!\n",
    "    while len(randomlist)!=150:\n",
    "        n=random.randint(0,len(rep_ID)-1)\n",
    "        if n not in randomlist:\n",
    "            randomlist.append(n)\n",
    "    return(randomlist)\n",
    "\n",
    "randomlist=randomize(rep_ID)\n",
    "\n",
    "def random_ID(rep_ID,randomlist):\n",
    "    PDB_ID=[]\n",
    "    for num in randomlist:\n",
    "        PDB_ID.append(rep_ID[int(num)])\n",
    "    return(PDB_ID)\n",
    "\n",
    "PDB_ID=random_ID(rep_ID,randomlist)\n",
    "\n",
    "print(len(PDB_ID))\n",
    "#print(PDB_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Making new directories !!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir 150_PDB_structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir rep_PDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I want to download from PDB all the PDB structures of the representative sequences (3532)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '4XXP'...\n",
      "Downloading PDB structure '4XXF'...\n",
      "Downloading PDB structure '4XXI'...\n",
      "Downloading PDB structure '4Y9I'...\n",
      "Downloading PDB structure '4XXB'...\n",
      "Downloading PDB structure '4Y94'...\n",
      "Downloading PDB structure '4Y8V'...\n",
      "Downloading PDB structure '4XWX'...\n",
      "Downloading PDB structure '4Y99'...\n",
      "Downloading PDB structure '4XYO'...\n",
      "Downloading PDB structure '4XYP'...\n",
      "Downloading PDB structure '4XXT'...\n",
      "Downloading PDB structure '4XY6'...\n",
      "Downloading PDB structure '4XXW'...\n",
      "Downloading PDB structure '4XZR'...\n",
      "Downloading PDB structure '4XZG'...\n",
      "Downloading PDB structure '4XZ9'...\n",
      "Downloading PDB structure '4XZX'...\n",
      "Downloading PDB structure '5A53'...\n",
      "Downloading PDB structure '5A4U'...\n",
      "Downloading PDB structure '5A4K'...\n",
      "Downloading PDB structure '5A4N'...\n",
      "Downloading PDB structure '5A5Y'...\n",
      "Downloading PDB structure '5A62'...\n",
      "Downloading PDB structure '5A6S'...\n",
      "Downloading PDB structure '5A6W'...\n",
      "Downloading PDB structure '5A8I'...\n",
      "Downloading PDB structure '5A8A'...\n",
      "Downloading PDB structure '5A7Z'...\n",
      "Downloading PDB structure '5A7L'...\n",
      "Downloading PDB structure '5A9H'...\n",
      "Downloading PDB structure '5A9D'...\n",
      "Downloading PDB structure '5A9C'...\n",
      "Downloading PDB structure '5A96'...\n",
      "Downloading PDB structure '5A98'...\n",
      "Downloading PDB structure '5ABS'...\n",
      "Downloading PDB structure '5ABR'...\n",
      "Downloading PDB structure '5ABD'...\n",
      "Downloading PDB structure '5MAB'...\n",
      "Downloading PDB structure '5MBG'...\n",
      "Downloading PDB structure '5MAL'...\n",
      "Downloading PDB structure '5MAO'...\n",
      "Downloading PDB structure '5MBU'...\n",
      "Downloading PDB structure '5MC7'...\n",
      "Downloading PDB structure '5MC9'...\n",
      "Downloading PDB structure '5MDI'...\n",
      "Downloading PDB structure '5MEB'...\n",
      "Downloading PDB structure '5MEC'...\n",
      "Downloading PDB structure '5MDU'...\n",
      "Downloading PDB structure '5MDT'...\n",
      "Downloading PDB structure '5ME5'...\n",
      "Downloading PDB structure '5ME7'...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-176-fc4a0e1040a6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m#Now I can downoad the PDB files\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mpdbl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPDBList\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0mpdbl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdownload_pdb_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPDB_ID_NOchain\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpdir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"rep_PDB/\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mfile_format\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"pdb\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/Bio/PDB/PDBList.py\u001b[0m in \u001b[0;36mdownload_pdb_files\u001b[0;34m(self, pdb_codes, obsolete, pdir, file_format, overwrite)\u001b[0m\n\u001b[1;32m    441\u001b[0m                 \u001b[0mpdir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpdir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m                 \u001b[0mfile_format\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfile_format\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 443\u001b[0;31m                 \u001b[0moverwrite\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moverwrite\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    444\u001b[0m             )\n\u001b[1;32m    445\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/Bio/PDB/PDBList.py\u001b[0m in \u001b[0;36mretrieve_pdb_file\u001b[0;34m(self, pdb_code, obsolete, pdir, file_format, overwrite)\u001b[0m\n\u001b[1;32m    336\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    337\u001b[0m             \u001b[0murlcleanup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 338\u001b[0;31m             \u001b[0murlretrieve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    339\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    340\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Desired structure doesn't exists\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36murlretrieve\u001b[0;34m(url, filename, reporthook, data)\u001b[0m\n\u001b[1;32m    245\u001b[0m     \u001b[0murl_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msplittype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 247\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mcontextlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclosing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    248\u001b[0m         \u001b[0mheaders\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    220\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    221\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 222\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    223\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0minstall_opener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopener\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    523\u001b[0m             \u001b[0mreq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 525\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    526\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    527\u001b[0m         \u001b[0;31m# post-process response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36m_open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    541\u001b[0m         \u001b[0mprotocol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    542\u001b[0m         result = self._call_chain(self.handle_open, protocol, protocol +\n\u001b[0;32m--> 543\u001b[0;31m                                   '_open', req)\n\u001b[0m\u001b[1;32m    544\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    545\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    501\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    502\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 503\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    504\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    505\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36mftp_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1540\u001b[0m                    \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'i'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'I'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'd'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'D'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1541\u001b[0m                     \u001b[0mtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1542\u001b[0;31m             \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mretrlen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfw\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrfile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1543\u001b[0m             \u001b[0mheaders\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1544\u001b[0m             \u001b[0mmtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmimetypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguess_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/urllib/request.py\u001b[0m in \u001b[0;36mretrfile\u001b[0;34m(self, file, type)\u001b[0m\n\u001b[1;32m   2395\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'TYPE '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0misdir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2396\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2397\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mftp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvoidcmd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2398\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mftplib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_errors\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2399\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/ftplib.py\u001b[0m in \u001b[0;36mvoidcmd\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    276\u001b[0m         \u001b[0;34m\"\"\"Send a command and expect a response beginning with '2'.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mputcmd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 278\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvoidresp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    279\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0msendport\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/ftplib.py\u001b[0m in \u001b[0;36mvoidresp\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    249\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mvoidresp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    250\u001b[0m         \u001b[0;34m\"\"\"Expect a response beginning with '2'.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 251\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    252\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'2'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    253\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0merror_reply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/ftplib.py\u001b[0m in \u001b[0;36mgetresp\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0;31m# Raise various errors if the response indicates an error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    235\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mgetresp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 236\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetmultiline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    237\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebugging\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    238\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'*resp*'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msanitize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/ftplib.py\u001b[0m in \u001b[0;36mgetmultiline\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    220\u001b[0m     \u001b[0;31m# these are separated by '\\n' characters in the string\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    221\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mgetmultiline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 222\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    223\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m             \u001b[0mcode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/ftplib.py\u001b[0m in \u001b[0;36mgetline\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    202\u001b[0m     \u001b[0;31m# Raise EOFError if the connection is closed\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mgetline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 204\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaxline\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaxline\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    206\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"got more than %d bytes\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaxline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    587\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    588\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 589\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    590\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    591\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#from Bio.PDB.PDBList import *\n",
    "PDB_ID_NOchain=[]\n",
    "for ID in rep_ID:\n",
    "    if ID[0:4] not in PDB_ID_NOchain:\n",
    "        PDB_ID_NOchain.append(ID[0:4])\n",
    "#Now I can downoad the PDB files\n",
    "pdbl=PDBList()\n",
    "pdbl.download_pdb_files(PDB_ID_NOchain,pdir=\"rep_PDB/\",file_format=\"pdb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DSSP Generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that I've downloaded all the PDB stuctures I want to perform the DSSP on them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I want now to save in the rep_ID only those files for which I have also the BDP downloaded in the folder rep_PDB (that contains 3336 PDB files).\n",
    "Because some of them are too big and I could not download them from PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3518\n"
     ]
    }
   ],
   "source": [
    "list_rep_PDB=[]\n",
    "rep_PDB=[]\n",
    "for filename in os.listdir(\"rep_PDB\"):\n",
    "    list_rep_PDB.append(filename[3:7].upper())\n",
    "#print(len(list_rep_PDB))\n",
    "for ID in rep_ID:\n",
    "    if ID[0:4] in list_rep_PDB:\n",
    "        rep_PDB.append(ID)\n",
    "print(len(rep_PDB))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rep_PDB' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-1e9d516675f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     19\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mran\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m \u001b[0mID\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrep_PDB\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0mID_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0mdssp_df2\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ID\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Chain\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"DSSP\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"Primary Structure\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rep_PDB' is not defined"
     ]
    }
   ],
   "source": [
    "#dssp_df=pd.DataFrame(columns=[\"ID\",\"Chain\",\"DSSP\",\"Primary Structure\",\"len\"])\n",
    "def convert(el):\n",
    "     if el == \"H\" or el ==\"G\" or el ==\"I\":\n",
    "        return \"H\"\n",
    "     elif el ==\"B\" or el ==\"E\":\n",
    "        return \"E\"\n",
    "     elif el==\"T\" or el ==\"S\" or el ==\"-\":\n",
    "        return \"C\"\n",
    "\n",
    "def sub_lower(aa):\n",
    "    if aa.islower():\n",
    "        return \"C\"\n",
    "    else:\n",
    "        return aa\n",
    "\n",
    "def random(rep_PDB):\n",
    "    import random\n",
    "    ran=random.sample(rep_PDB,1)[0]\n",
    "    return(ran)\n",
    "\n",
    "ID=random(rep_PDB)\n",
    "ID_list=[]\n",
    "dssp_df2=pd.DataFrame(columns=[\"ID\",\"Chain\",\"DSSP\",\"Primary Structure\"])\n",
    "while len(dssp_df2)!=150: #Specify the number of sequeces that you want in your dataframe\n",
    "    if ID not in ID_list:\n",
    "        ID_list.append(ID)\n",
    "        prim_str=\"\"\n",
    "        sec_str=\"\"\n",
    "        #print(ID)\n",
    "        dssp_tuple=dssp_dict_from_pdb_file(\"rep_PDB/pdb\"+ID[0:4].lower()+\".ent\",DSSP=\"mkdssp\")\n",
    "        dssp_dict=dssp_tuple[0]\n",
    "        index=1\n",
    "        for line in dssp_dict:\n",
    "            if line[0]==ID[5]:\n",
    "                if index==1:\n",
    "                    temp_line=line[1][1]\n",
    "                    prim_str=prim_str+sub_lower(dssp_dict[line][0])\n",
    "                    sec_str=sec_str+convert(dssp_dict[line][1])\n",
    "                    index=index+1\n",
    "                else:\n",
    "                    actual_line=line[1][1]\n",
    "                    if actual_line-1==temp_line:\n",
    "                        prim_str=prim_str+sub_lower(dssp_dict[line][0])\n",
    "                        sec_str=sec_str+convert(dssp_dict[line][1])\n",
    "                    else:\n",
    "                        prim_str=prim_str+\"error\" #X\n",
    "                        sec_str=sec_str+\"C\"\n",
    "                    temp_line=line[1][1]\n",
    "        if \"error\" not in prim_str and len(prim_str)>50:\n",
    "            dssp_df2=dssp_df2.append({\"ID\":ID[0:4],\"Chain\":ID[5],\"DSSP\":sec_str,\"Primary Structure\":prim_str,\"len\":int(len(sec_str)) },ignore_index=True)\n",
    "    else:\n",
    "        ID=random(rep_PDB)\n",
    "        \n",
    "dssp_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Chain</th>\n",
       "      <th>DSSP</th>\n",
       "      <th>Primary Structure</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5L8X</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCEECCCCCCEEEECCCCCCCHHHHHHC...</td>\n",
       "      <td>SHMANKREPAPGWPIVSGEYVVGNPESCVGVVTLGSHGLEQACIDA...</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5WOQ</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCHHHHHHHHHHHHHHCCCCCHHHHHHHHCCCHHHHHHHHCCCCC...</td>\n",
       "      <td>TALLREVIGDVLRNARTDQGRTLREVSDAARVSLGYLSEVERGRKE...</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4ZPU</td>\n",
       "      <td>A</td>\n",
       "      <td>CCHHHHHHHHHCHHHHHHHHHHHHHHCCCCCCCCCCCEEEEEECCC...</td>\n",
       "      <td>MPPSLRKAVAAAIGGGAIAIASVLITGPSGNDGLEGVSYIPYKDIV...</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6GMR</td>\n",
       "      <td>B</td>\n",
       "      <td>CEEEEEEEEECEEEEEEEECCCEHHHHHHHHHHHHCCCHHHEEEEE...</td>\n",
       "      <td>MDVFLMIRRHKTTIFTDAKESSTVFELKRIVEGILKRPPDEQRLYK...</td>\n",
       "      <td>105.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5WSD</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCCEEEEEHHHCCCEEEEECCEEEEEEEEEECCCCCCCEEEEEEE...</td>\n",
       "      <td>GPSGMILKRAYDVTPQKISTDKVRGVRKRVLIGLKDAPNFVMRLFT...</td>\n",
       "      <td>118.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>6JL3</td>\n",
       "      <td>A</td>\n",
       "      <td>CCEEEEEEECCCEEEEEEECCCCEHHHHHHHHHCCCCCCHHHEEEE...</td>\n",
       "      <td>MVINVSFKVTGGKEFTVAIEPDITVLDLKKICAEHVDIPVEAQRII...</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>6DRE</td>\n",
       "      <td>B</td>\n",
       "      <td>CHHHHHHHHHHHHHHHHHHHHHCCCCHHHHHHHHHHCCHHHHHHHH...</td>\n",
       "      <td>EQAKVWTQTARANAEKNNAQLSTLLTDDQIGAIYGYTTNEGYTALN...</td>\n",
       "      <td>170.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>5K31</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCHHHHHHHHHHHHHHHHHHHHCCCCCCCCCECCHHHHHHHCCCC...</td>\n",
       "      <td>VRDRDLEVDTTLKSLSQQIENIRSPEGSRKNPARTCRDLKMCHSDW...</td>\n",
       "      <td>241.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>6S8X</td>\n",
       "      <td>A</td>\n",
       "      <td>CHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...</td>\n",
       "      <td>TAGYRSLTYEEVLQELVKHKELLRRKDTHIRELEDYIDNLLVRVME...</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>6U9J</td>\n",
       "      <td>A</td>\n",
       "      <td>CCHHHHHCCCCCCCHHHHHHHHCCCHHHHHHCCCCEEEEEHHHHHH...</td>\n",
       "      <td>VSLQEFLKTEPDGTLEVVAEQYNTTLLEVVRNLPSSTVVPGDKFDT...</td>\n",
       "      <td>160.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID Chain                                               DSSP  \\\n",
       "0    5L8X     A  CCCCCCCCCCCCCCCCCCCCEECCCCCCEEEECCCCCCCHHHHHHC...   \n",
       "1    5WOQ     A  CCCHHHHHHHHHHHHHHCCCCCHHHHHHHHCCCHHHHHHHHCCCCC...   \n",
       "2    4ZPU     A  CCHHHHHHHHHCHHHHHHHHHHHHHHCCCCCCCCCCCEEEEEECCC...   \n",
       "3    6GMR     B  CEEEEEEEEECEEEEEEEECCCEHHHHHHHHHHHHCCCHHHEEEEE...   \n",
       "4    5WSD     A  CCCCEEEEEHHHCCCEEEEECCEEEEEEEEEECCCCCCCEEEEEEE...   \n",
       "..    ...   ...                                                ...   \n",
       "145  6JL3     A  CCEEEEEEECCCEEEEEEECCCCEHHHHHHHHHCCCCCCHHHEEEE...   \n",
       "146  6DRE     B  CHHHHHHHHHHHHHHHHHHHHHCCCCHHHHHHHHHHCCHHHHHHHH...   \n",
       "147  5K31     A  CCCHHHHHHHHHHHHHHHHHHHHCCCCCCCCCECCHHHHHHHCCCC...   \n",
       "148  6S8X     A  CHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...   \n",
       "149  6U9J     A  CCHHHHHCCCCCCCHHHHHHHHCCCHHHHHHCCCCEEEEEHHHHHH...   \n",
       "\n",
       "                                     Primary Structure    len  \n",
       "0    SHMANKREPAPGWPIVSGEYVVGNPESCVGVVTLGSHGLEQACIDA...  172.0  \n",
       "1    TALLREVIGDVLRNARTDQGRTLREVSDAARVSLGYLSEVERGRKE...   80.0  \n",
       "2    MPPSLRKAVAAAIGGGAIAIASVLITGPSGNDGLEGVSYIPYKDIV...  165.0  \n",
       "3    MDVFLMIRRHKTTIFTDAKESSTVFELKRIVEGILKRPPDEQRLYK...  105.0  \n",
       "4    GPSGMILKRAYDVTPQKISTDKVRGVRKRVLIGLKDAPNFVMRLFT...  118.0  \n",
       "..                                                 ...    ...  \n",
       "145  MVINVSFKVTGGKEFTVAIEPDITVLDLKKICAEHVDIPVEAQRII...   74.0  \n",
       "146  EQAKVWTQTARANAEKNNAQLSTLLTDDQIGAIYGYTTNEGYTALN...  170.0  \n",
       "147  VRDRDLEVDTTLKSLSQQIENIRSPEGSRKNPARTCRDLKMCHSDW...  241.0  \n",
       "148  TAGYRSLTYEEVLQELVKHKELLRRKDTHIRELEDYIDNLLVRVME...   53.0  \n",
       "149  VSLQEFLKTEPDGTLEVVAEQYNTTLLEVVRNLPSSTVVPGDKFDT...  160.0  \n",
       "\n",
       "[150 rows x 5 columns]"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dssp_final=dssp_df2\n",
    "dssp_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Chain</th>\n",
       "      <th>DSSP</th>\n",
       "      <th>Primary Structure</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5L8X</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCCCCCCCCCCCCCCCCCCEECCCCCCEEEECCCCCCCHHHHHHC...</td>\n",
       "      <td>SHMANKREPAPGWPIVSGEYVVGNPESCVGVVTLGSHGLEQACIDA...</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5WOQ</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCHHHHHHHHHHHHHHCCCCCHHHHHHHHCCCHHHHHHHHCCCCC...</td>\n",
       "      <td>TALLREVIGDVLRNARTDQGRTLREVSDAARVSLGYLSEVERGRKE...</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4ZPU</td>\n",
       "      <td>A</td>\n",
       "      <td>CCHHHHHHHHHCHHHHHHHHHHHHHHCCCCCCCCCCCEEEEEECCC...</td>\n",
       "      <td>MPPSLRKAVAAAIGGGAIAIASVLITGPSGNDGLEGVSYIPYKDIV...</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6GMR</td>\n",
       "      <td>B</td>\n",
       "      <td>CEEEEEEEEECEEEEEEEECCCEHHHHHHHHHHHHCCCHHHEEEEE...</td>\n",
       "      <td>MDVFLMIRRHKTTIFTDAKESSTVFELKRIVEGILKRPPDEQRLYK...</td>\n",
       "      <td>105.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5WSD</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCCEEEEEHHHCCCEEEEECCEEEEEEEEEECCCCCCCEEEEEEE...</td>\n",
       "      <td>GPSGMILKRAYDVTPQKISTDKVRGVRKRVLIGLKDAPNFVMRLFT...</td>\n",
       "      <td>118.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>6JL3</td>\n",
       "      <td>A</td>\n",
       "      <td>CCEEEEEEECCCEEEEEEECCCCEHHHHHHHHHCCCCCCHHHEEEE...</td>\n",
       "      <td>MVINVSFKVTGGKEFTVAIEPDITVLDLKKICAEHVDIPVEAQRII...</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>6DRE</td>\n",
       "      <td>B</td>\n",
       "      <td>CHHHHHHHHHHHHHHHHHHHHHCCCCHHHHHHHHHHCCHHHHHHHH...</td>\n",
       "      <td>EQAKVWTQTARANAEKNNAQLSTLLTDDQIGAIYGYTTNEGYTALN...</td>\n",
       "      <td>170.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>5K31</td>\n",
       "      <td>A</td>\n",
       "      <td>CCCHHHHHHHHHHHHHHHHHHHHCCCCCCCCCECCHHHHHHHCCCC...</td>\n",
       "      <td>VRDRDLEVDTTLKSLSQQIENIRSPEGSRKNPARTCRDLKMCHSDW...</td>\n",
       "      <td>241.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>6S8X</td>\n",
       "      <td>A</td>\n",
       "      <td>CHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...</td>\n",
       "      <td>TAGYRSLTYEEVLQELVKHKELLRRKDTHIRELEDYIDNLLVRVME...</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>6U9J</td>\n",
       "      <td>A</td>\n",
       "      <td>CCHHHHHCCCCCCCHHHHHHHHCCCHHHHHHCCCCEEEEEHHHHHH...</td>\n",
       "      <td>VSLQEFLKTEPDGTLEVVAEQYNTTLLEVVRNLPSSTVVPGDKFDT...</td>\n",
       "      <td>160.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID Chain                                               DSSP  \\\n",
       "0    5L8X     A  CCCCCCCCCCCCCCCCCCCCEECCCCCCEEEECCCCCCCHHHHHHC...   \n",
       "1    5WOQ     A  CCCHHHHHHHHHHHHHHCCCCCHHHHHHHHCCCHHHHHHHHCCCCC...   \n",
       "2    4ZPU     A  CCHHHHHHHHHCHHHHHHHHHHHHHHCCCCCCCCCCCEEEEEECCC...   \n",
       "3    6GMR     B  CEEEEEEEEECEEEEEEEECCCEHHHHHHHHHHHHCCCHHHEEEEE...   \n",
       "4    5WSD     A  CCCCEEEEEHHHCCCEEEEECCEEEEEEEEEECCCCCCCEEEEEEE...   \n",
       "..    ...   ...                                                ...   \n",
       "145  6JL3     A  CCEEEEEEECCCEEEEEEECCCCEHHHHHHHHHCCCCCCHHHEEEE...   \n",
       "146  6DRE     B  CHHHHHHHHHHHHHHHHHHHHHCCCCHHHHHHHHHHCCHHHHHHHH...   \n",
       "147  5K31     A  CCCHHHHHHHHHHHHHHHHHHHHCCCCCCCCCECCHHHHHHHCCCC...   \n",
       "148  6S8X     A  CHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...   \n",
       "149  6U9J     A  CCHHHHHCCCCCCCHHHHHHHHCCCHHHHHHCCCCEEEEEHHHHHH...   \n",
       "\n",
       "                                     Primary Structure    len  \n",
       "0    SHMANKREPAPGWPIVSGEYVVGNPESCVGVVTLGSHGLEQACIDA...  172.0  \n",
       "1    TALLREVIGDVLRNARTDQGRTLREVSDAARVSLGYLSEVERGRKE...   80.0  \n",
       "2    MPPSLRKAVAAAIGGGAIAIASVLITGPSGNDGLEGVSYIPYKDIV...  165.0  \n",
       "3    MDVFLMIRRHKTTIFTDAKESSTVFELKRIVEGILKRPPDEQRLYK...  105.0  \n",
       "4    GPSGMILKRAYDVTPQKISTDKVRGVRKRVLIGLKDAPNFVMRLFT...  118.0  \n",
       "..                                                 ...    ...  \n",
       "145  MVINVSFKVTGGKEFTVAIEPDITVLDLKKICAEHVDIPVEAQRII...   74.0  \n",
       "146  EQAKVWTQTARANAEKNNAQLSTLLTDDQIGAIYGYTTNEGYTALN...  170.0  \n",
       "147  VRDRDLEVDTTLKSLSQQIENIRSPEGSRKNPARTCRDLKMCHSDW...  241.0  \n",
       "148  TAGYRSLTYEEVLQELVKHKELLRRKDTHIRELEDYIDNLLVRVME...   53.0  \n",
       "149  VSLQEFLKTEPDGTLEVVAEQYNTTLLEVVRNLPSSTVVPGDKFDT...  160.0  \n",
       "\n",
       "[150 rows x 5 columns]"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dssp_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I want to check if the percentage of the secondary structures (C-coil, E-beta sheet, H-helix) in the DSSP are more or less balanced."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentage of  C :  36.504352110362944 %\n",
      "Percentage of  H :  40.396616850057484 %\n",
      "Percentage of  E :  23.09903103957957 %\n"
     ]
    }
   ],
   "source": [
    "dict={}\n",
    "tot=0\n",
    "for dssp in os.listdir(\"Blindset_DSSP\"):\n",
    "    dssp=open(\"Blindset_DSSP/\"+dssp)\n",
    "    for line in dssp:\n",
    "        if line.startswith(\">\"):\n",
    "            dssp=dssp.readline().rstrip()\n",
    "            for el in dssp:\n",
    "                key=dict.keys()\n",
    "                if el in key:\n",
    "                    dict[el]+=1\n",
    "                else:\n",
    "                    dict[el]=1\n",
    "for key in dict:\n",
    "    tot=tot+dict[key]\n",
    "for key in dict:\n",
    "    print(\"Percentage of \" , key,\": \", (dict[key]/tot)*100,\"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I can easily make a cake graph to put in the report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving the DSSP and FASTA files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I want to save all the 150 DSSP and fasta in separate files in separate folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Blindset_Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Blindset_DSSP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index in dssp_final.index:\n",
    "    ID=\">\"+dssp_final.loc[index,\"ID\"]+\"_\"+dssp_final.loc[index,\"Chain\"]\n",
    "    fasta=dssp_final.loc[index,\"Primary Structure\"]\n",
    "    file=open(\"Blindset_Fasta/\"+ID[1:]+\".fasta\",\"w+\")\n",
    "    file.write(ID)\n",
    "    file.write(\"\\n\")\n",
    "    file.write(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [],
   "source": [
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index in dssp_final.index:\n",
    "    ID=\">\"+dssp_final.loc[index,\"ID\"]+\"_\"+dssp_final.loc[index,\"Chain\"]\n",
    "    DSSP=dssp_final.loc[index,\"DSSP\"]\n",
    "    file=open(\"Blindset_DSSP/\"+ID[1:]+\".dssp\",\"w+\")\n",
    "    file.write(ID)\n",
    "    file.write(\"\\n\")\n",
    "    file.write(DSSP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [],
   "source": [
    "file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running PSI-BLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First I have to convert the \"uniprot_sprot.fasta\" file in a database so that I can then blast against it the fasta sequences of the Blind set (test set)/JPred4 (training set) sets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 09/25/2020 19:07:29\n",
      "New DB name:   /mnt/EA7C9C107C9BD5A3/User/BIOINFORMATICA.BOLOGNA/LAB_2/Project/Project2/uniprot_sprot.fasta\n",
      "New DB title:  uniprot_sprot.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /mnt/EA7C9C107C9BD5A3/User/BIOINFORMATICA.BOLOGNA/LAB_2/Project/Project2/uniprot_sprot.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 563082 sequences in 34.8935 seconds.\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in uniprot_sprot.fasta -dbtype prot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I can run psi-blast.\n",
    "Documentation: https://biopython.org/docs/1.75/api/Bio.Blast.Applications.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Psiblast_alns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Psiblast_pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from Bio.Blast.Applications import NcbipsiblastCommandline\n",
    "for fasta_file in os.listdir(\"Blindset_Fasta/\"):\n",
    "    cline=NcbipsiblastCommandline(query=\"Blindset_Fasta/\"+fasta_file,\n",
    "                                 db='uniprot_sprot.fasta',\n",
    "                                 evalue= 0.01,\n",
    "                                 num_iterations=3,\n",
    "                                 out=\"Psiblast_alns/\"+fasta_file[0:6]+\".alns.blast\",\n",
    "                                 out_ascii_pssm=\"Psiblast_pssm//\"+fasta_file[0:6]+\".pssm\", #I could put another directory for the output\n",
    "                                 num_descriptions=10000,\n",
    "                                 num_alignments=10000,\n",
    "                                 num_threads=4)\n",
    "    cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Psiblast_alns_JPred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Psiblast_pssm_JPred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fasta_file in os.listdir(\"fasta_Jpred/\"):\n",
    "    base_name=os.path.basename(fasta_file)\n",
    "    name=os.path.splitext(base_name)[0]\n",
    "    cline=NcbipsiblastCommandline(query=\"fasta_Jpred/\"+base_name,\n",
    "                                 db='uniprot_sprot.fasta',\n",
    "                                 evalue= 0.01,\n",
    "                                 num_iterations=3,\n",
    "                                 out=\"Psiblast_alns_JPred/\"+name+\".alns.blast\",\n",
    "                                 out_ascii_pssm=\"Psiblast_pssm_JPred/\"+name+\".pssm\", #I could put another directory for the output\n",
    "                                 num_descriptions=10000,\n",
    "                                 num_alignments=10000,\n",
    "                                 num_threads=4)\n",
    "    cline()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract the Profile matrices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I want to extract only the Profile matrices from the .pssm files and save the pssm matrix in the new folder. I have to normalize (by dividing by 100) all the value in the matrxi and delete all those those profile that have a matrix composide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir Blindset_pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This  1  PSSM has a matrxi of 0s only:  5ABV_B\n",
      "This  2  PSSM has a matrxi of 0s only:  5VDB_A\n",
      "This  3  PSSM has a matrxi of 0s only:  5MU3_B\n",
      "This  4  PSSM has a matrxi of 0s only:  6NNW_A\n",
      "This  5  PSSM has a matrxi of 0s only:  6B4H_B\n"
     ]
    }
   ],
   "source": [
    "def pssm_fun(pssm):\n",
    "    pssm_list=[]\n",
    "    tot=0\n",
    "    for line in pssm:\n",
    "        if \"Last\" in line:\n",
    "            colname=pssm.readline().split()[20:]\n",
    "            for line in pssm:\n",
    "                line_list=[]\n",
    "                line_list=line.rstrip().split()[22:42]\n",
    "                if line_list==[]:\n",
    "                    break\n",
    "                else:\n",
    "                    for el in range(len(line_list)):\n",
    "                        line_list[el]=int(line_list[el])/100\n",
    "                        tot=tot+float(line_list[el])\n",
    "                    pssm_list.append(line_list)\n",
    "    pssm_df=pd.DataFrame(pssm_list,columns=colname)\n",
    "    pssm_df.index=pssm_df.index+1\n",
    "    return(pssm_df,tot)\n",
    "\n",
    "count=0\n",
    "for pssm in os.listdir(\"Psiblast_pssm\"):\n",
    "    if pssm.endswith(\".pssm\"):\n",
    "        pssm_file=open(\"Psiblast_pssm/\"+pssm)\n",
    "        pssm_name=pssm.split(\".\")\n",
    "        pssm_name=\".\".join(pssm_name[:-1])\n",
    "        pssm_df,tot=pssm_fun(pssm_file)\n",
    "        #pssm_df\n",
    "        if tot==0:\n",
    "            count+=1\n",
    "            print(\"This \", count,\"° PSSM has a matrxi of 0s only: \" ,pssm_name)\n",
    "        else:\n",
    "            pssm_df.to_csv(\"Blindset_pssm/\"+pssm_name+\".pssm\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir JPred_pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This  1  PSSM has a matrxi of 0s only:  d1v74b_\n",
      "This  2  PSSM has a matrxi of 0s only:  d1vk5a_\n",
      "This  3  PSSM has a matrxi of 0s only:  d1w53a_\n",
      "This  4  PSSM has a matrxi of 0s only:  d3h7ia2\n",
      "This  5  PSSM has a matrxi of 0s only:  d2ebfx1\n",
      "This  6  PSSM has a matrxi of 0s only:  d2enda_\n",
      "This  7  PSSM has a matrxi of 0s only:  d2erla_\n",
      "This  8  PSSM has a matrxi of 0s only:  d2euca1\n",
      "This  9  PSSM has a matrxi of 0s only:  d2f2ha1\n",
      "This  10  PSSM has a matrxi of 0s only:  d2gsva1\n",
      "This  11  PSSM has a matrxi of 0s only:  d1ryla_\n",
      "This  12  PSSM has a matrxi of 0s only:  d1yoza1\n",
      "This  13  PSSM has a matrxi of 0s only:  d1yu0a1\n",
      "This  14  PSSM has a matrxi of 0s only:  d1s7za_\n",
      "This  15  PSSM has a matrxi of 0s only:  d1seda_\n",
      "This  16  PSSM has a matrxi of 0s only:  d1sf9a_\n",
      "This  17  PSSM has a matrxi of 0s only:  d1pv5a_\n",
      "This  18  PSSM has a matrxi of 0s only:  d1e7la1\n",
      "This  19  PSSM has a matrxi of 0s only:  d1el6a_\n",
      "This  20  PSSM has a matrxi of 0s only:  d1ezga_\n",
      "This  21  PSSM has a matrxi of 0s only:  d1dl5a2\n",
      "This  22  PSSM has a matrxi of 0s only:  d1dtdb_\n",
      "This  23  PSSM has a matrxi of 0s only:  d2wkxa1\n",
      "This  24  PSSM has a matrxi of 0s only:  d1nrja_\n",
      "This  25  PSSM has a matrxi of 0s only:  d2ffga1\n",
      "This  26  PSSM has a matrxi of 0s only:  d2fpna1\n",
      "This  27  PSSM has a matrxi of 0s only:  d2ftxa1\n",
      "This  28  PSSM has a matrxi of 0s only:  d2g3ra1\n",
      "This  29  PSSM has a matrxi of 0s only:  d1zkea1\n",
      "This  30  PSSM has a matrxi of 0s only:  d2b59b2\n",
      "This  31  PSSM has a matrxi of 0s only:  d1kafa_\n",
      "This  32  PSSM has a matrxi of 0s only:  d1ti6b1\n",
      "This  33  PSSM has a matrxi of 0s only:  d2ohwa1\n",
      "This  34  PSSM has a matrxi of 0s only:  d1o75a1\n",
      "This  35  PSSM has a matrxi of 0s only:  d2pmra1\n",
      "This  36  PSSM has a matrxi of 0s only:  d2prva1\n",
      "This  37  PSSM has a matrxi of 0s only:  d2psba1\n",
      "This  38  PSSM has a matrxi of 0s only:  d2bhua2\n",
      "This  39  PSSM has a matrxi of 0s only:  d1kp6a_\n",
      "This  40  PSSM has a matrxi of 0s only:  d1kpta_\n",
      "This  41  PSSM has a matrxi of 0s only:  d1m1ha1\n",
      "This  42  PSSM has a matrxi of 0s only:  d1n7za_\n",
      "This  43  PSSM has a matrxi of 0s only:  d2choa1\n",
      "This  44  PSSM has a matrxi of 0s only:  d2d0oa1\n",
      "This  45  PSSM has a matrxi of 0s only:  d2dlba1\n",
      "This  46  PSSM has a matrxi of 0s only:  d1ugia_\n",
      "This  47  PSSM has a matrxi of 0s only:  d1uoya_\n",
      "This  48  PSSM has a matrxi of 0s only:  d2rdea1\n"
     ]
    }
   ],
   "source": [
    "def pssm_fun(pssm):\n",
    "    pssm_list=[]\n",
    "    tot=0\n",
    "    for line in pssm:\n",
    "        if \"Last\" in line:\n",
    "            colname=pssm.readline().split()[20:]\n",
    "            for line in pssm:\n",
    "                line_list=[]\n",
    "                line_list=line.rstrip().split()[22:42]\n",
    "                if line_list==[]:\n",
    "                    break\n",
    "                else:\n",
    "                    for el in range(len(line_list)):\n",
    "                        line_list[el]=int(line_list[el])/100\n",
    "                        tot=tot+float(line_list[el])\n",
    "                    pssm_list.append(line_list)\n",
    "    pssm_df=pd.DataFrame(pssm_list,columns=colname)\n",
    "    pssm_df.index=pssm_df.index+1\n",
    "    return(pssm_df,tot)\n",
    "\n",
    "count=0\n",
    "for pssm in os.listdir(\"Psiblast_pssm_JPred\"):\n",
    "    if pssm.endswith(\".pssm\"):\n",
    "        pssm_file=open(\"Psiblast_pssm_JPred/\"+pssm)\n",
    "        pssm_name=pssm.split(\".\")\n",
    "        pssm_name=\".\".join(pssm_name[:-1])\n",
    "        pssm_df,tot=pssm_fun(pssm_file) #function recall\n",
    "        #print(pssm_df)\n",
    "        if tot==0:\n",
    "            count+=1\n",
    "            print(\"This \", count,\" PSSM has a matrxi of 0s only: \" ,pssm_name)\n",
    "        else:\n",
    "            pssm_df.to_csv(\"JPred_pssm/\"+pssm_name+\".pssm\",sep=\"\\t\")\n",
    "#pssm_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Develop the Gor-training program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **GOR method** is based on probability parameters derived from empirical studies of known protein tertiary structures solved by X-ray crystallography. However, unlike the Chou-Fasman, the Gor method takes into account not only the propensity of individual amino acids to form particular secondary structures, but also the conditional probability of the amino acid to form a secondary structure given that its immediate neighbours have already formed that structure. The methods is essentially **Bayesian** in its analysis. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have to pick each profile matrix that we have generated and extracted from PsiBlast and generate the gor matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Put here the path of the folder with the profiles and the secondary structure of the sequences for which you want to do the training :**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "pssm_path=\"JPred_pssm/\" #Profile path\n",
    "dssp_path=\"JPred_dssp/\" #Secondary structure path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of the secondary structures:  {'C': 84297, 'H': 70534, 'E': 43972}\n",
      "CPU times: user 5.84 s, sys: 181 ms, total: 6.02 s\n",
      "Wall time: 9.37 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>R</th>\n",
       "      <th>N</th>\n",
       "      <th>D</th>\n",
       "      <th>C</th>\n",
       "      <th>Q</th>\n",
       "      <th>E</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>K</th>\n",
       "      <th>M</th>\n",
       "      <th>F</th>\n",
       "      <th>P</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>V</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>R-8_H</th>\n",
       "      <td>0.031335</td>\n",
       "      <td>0.018633</td>\n",
       "      <td>0.014761</td>\n",
       "      <td>0.021476</td>\n",
       "      <td>0.004575</td>\n",
       "      <td>0.014522</td>\n",
       "      <td>0.026603</td>\n",
       "      <td>0.023597</td>\n",
       "      <td>0.008305</td>\n",
       "      <td>0.020903</td>\n",
       "      <td>0.035207</td>\n",
       "      <td>0.021388</td>\n",
       "      <td>0.008456</td>\n",
       "      <td>0.013840</td>\n",
       "      <td>0.015724</td>\n",
       "      <td>0.022866</td>\n",
       "      <td>0.019575</td>\n",
       "      <td>0.004304</td>\n",
       "      <td>0.011312</td>\n",
       "      <td>0.023634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-7_H</th>\n",
       "      <td>0.032013</td>\n",
       "      <td>0.018729</td>\n",
       "      <td>0.014715</td>\n",
       "      <td>0.021577</td>\n",
       "      <td>0.004574</td>\n",
       "      <td>0.014702</td>\n",
       "      <td>0.027173</td>\n",
       "      <td>0.023042</td>\n",
       "      <td>0.008143</td>\n",
       "      <td>0.020807</td>\n",
       "      <td>0.035373</td>\n",
       "      <td>0.021351</td>\n",
       "      <td>0.008483</td>\n",
       "      <td>0.013792</td>\n",
       "      <td>0.015211</td>\n",
       "      <td>0.022843</td>\n",
       "      <td>0.019518</td>\n",
       "      <td>0.004285</td>\n",
       "      <td>0.011227</td>\n",
       "      <td>0.023504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-6_H</th>\n",
       "      <td>0.032711</td>\n",
       "      <td>0.018709</td>\n",
       "      <td>0.014504</td>\n",
       "      <td>0.021578</td>\n",
       "      <td>0.004625</td>\n",
       "      <td>0.014919</td>\n",
       "      <td>0.027531</td>\n",
       "      <td>0.022268</td>\n",
       "      <td>0.008049</td>\n",
       "      <td>0.021052</td>\n",
       "      <td>0.036033</td>\n",
       "      <td>0.021327</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0.013814</td>\n",
       "      <td>0.014686</td>\n",
       "      <td>0.022586</td>\n",
       "      <td>0.019409</td>\n",
       "      <td>0.004293</td>\n",
       "      <td>0.011017</td>\n",
       "      <td>0.023378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-5_H</th>\n",
       "      <td>0.033496</td>\n",
       "      <td>0.018969</td>\n",
       "      <td>0.014333</td>\n",
       "      <td>0.021467</td>\n",
       "      <td>0.004577</td>\n",
       "      <td>0.015187</td>\n",
       "      <td>0.028119</td>\n",
       "      <td>0.021410</td>\n",
       "      <td>0.007920</td>\n",
       "      <td>0.021205</td>\n",
       "      <td>0.036702</td>\n",
       "      <td>0.021536</td>\n",
       "      <td>0.008618</td>\n",
       "      <td>0.013613</td>\n",
       "      <td>0.014068</td>\n",
       "      <td>0.022236</td>\n",
       "      <td>0.019130</td>\n",
       "      <td>0.004277</td>\n",
       "      <td>0.010902</td>\n",
       "      <td>0.023181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-4_H</th>\n",
       "      <td>0.034159</td>\n",
       "      <td>0.019174</td>\n",
       "      <td>0.014201</td>\n",
       "      <td>0.021362</td>\n",
       "      <td>0.004424</td>\n",
       "      <td>0.015591</td>\n",
       "      <td>0.028883</td>\n",
       "      <td>0.020069</td>\n",
       "      <td>0.007799</td>\n",
       "      <td>0.021404</td>\n",
       "      <td>0.037283</td>\n",
       "      <td>0.021816</td>\n",
       "      <td>0.008687</td>\n",
       "      <td>0.013470</td>\n",
       "      <td>0.013333</td>\n",
       "      <td>0.021942</td>\n",
       "      <td>0.018862</td>\n",
       "      <td>0.004315</td>\n",
       "      <td>0.010845</td>\n",
       "      <td>0.023028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R4_tot</th>\n",
       "      <td>0.081727</td>\n",
       "      <td>0.051768</td>\n",
       "      <td>0.041783</td>\n",
       "      <td>0.057616</td>\n",
       "      <td>0.014798</td>\n",
       "      <td>0.038320</td>\n",
       "      <td>0.067987</td>\n",
       "      <td>0.071549</td>\n",
       "      <td>0.022834</td>\n",
       "      <td>0.060092</td>\n",
       "      <td>0.095151</td>\n",
       "      <td>0.059887</td>\n",
       "      <td>0.020857</td>\n",
       "      <td>0.039542</td>\n",
       "      <td>0.043656</td>\n",
       "      <td>0.061267</td>\n",
       "      <td>0.053969</td>\n",
       "      <td>0.013180</td>\n",
       "      <td>0.033308</td>\n",
       "      <td>0.070707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R5_tot</th>\n",
       "      <td>0.081802</td>\n",
       "      <td>0.051783</td>\n",
       "      <td>0.041794</td>\n",
       "      <td>0.057669</td>\n",
       "      <td>0.014804</td>\n",
       "      <td>0.038295</td>\n",
       "      <td>0.067940</td>\n",
       "      <td>0.071696</td>\n",
       "      <td>0.022848</td>\n",
       "      <td>0.060033</td>\n",
       "      <td>0.095121</td>\n",
       "      <td>0.059897</td>\n",
       "      <td>0.020861</td>\n",
       "      <td>0.039552</td>\n",
       "      <td>0.043544</td>\n",
       "      <td>0.061239</td>\n",
       "      <td>0.053944</td>\n",
       "      <td>0.013191</td>\n",
       "      <td>0.033342</td>\n",
       "      <td>0.070646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R6_tot</th>\n",
       "      <td>0.081889</td>\n",
       "      <td>0.051763</td>\n",
       "      <td>0.041819</td>\n",
       "      <td>0.057704</td>\n",
       "      <td>0.014796</td>\n",
       "      <td>0.038281</td>\n",
       "      <td>0.067898</td>\n",
       "      <td>0.071829</td>\n",
       "      <td>0.022907</td>\n",
       "      <td>0.059915</td>\n",
       "      <td>0.095113</td>\n",
       "      <td>0.059874</td>\n",
       "      <td>0.020858</td>\n",
       "      <td>0.039556</td>\n",
       "      <td>0.043509</td>\n",
       "      <td>0.061282</td>\n",
       "      <td>0.053904</td>\n",
       "      <td>0.013197</td>\n",
       "      <td>0.033330</td>\n",
       "      <td>0.070576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R7_tot</th>\n",
       "      <td>0.081975</td>\n",
       "      <td>0.051712</td>\n",
       "      <td>0.041873</td>\n",
       "      <td>0.057743</td>\n",
       "      <td>0.014778</td>\n",
       "      <td>0.038265</td>\n",
       "      <td>0.067881</td>\n",
       "      <td>0.071950</td>\n",
       "      <td>0.022905</td>\n",
       "      <td>0.059925</td>\n",
       "      <td>0.095047</td>\n",
       "      <td>0.059776</td>\n",
       "      <td>0.020869</td>\n",
       "      <td>0.039515</td>\n",
       "      <td>0.043529</td>\n",
       "      <td>0.061289</td>\n",
       "      <td>0.053884</td>\n",
       "      <td>0.013182</td>\n",
       "      <td>0.033347</td>\n",
       "      <td>0.070555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R8_tot</th>\n",
       "      <td>0.082043</td>\n",
       "      <td>0.051672</td>\n",
       "      <td>0.041885</td>\n",
       "      <td>0.057779</td>\n",
       "      <td>0.014747</td>\n",
       "      <td>0.038225</td>\n",
       "      <td>0.067933</td>\n",
       "      <td>0.072053</td>\n",
       "      <td>0.022922</td>\n",
       "      <td>0.059917</td>\n",
       "      <td>0.094951</td>\n",
       "      <td>0.059744</td>\n",
       "      <td>0.020856</td>\n",
       "      <td>0.039534</td>\n",
       "      <td>0.043510</td>\n",
       "      <td>0.061301</td>\n",
       "      <td>0.053866</td>\n",
       "      <td>0.013150</td>\n",
       "      <td>0.033359</td>\n",
       "      <td>0.070552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A         R         N         D         C         Q         E  \\\n",
       "R-8_H   0.031335  0.018633  0.014761  0.021476  0.004575  0.014522  0.026603   \n",
       "R-7_H   0.032013  0.018729  0.014715  0.021577  0.004574  0.014702  0.027173   \n",
       "R-6_H   0.032711  0.018709  0.014504  0.021578  0.004625  0.014919  0.027531   \n",
       "R-5_H   0.033496  0.018969  0.014333  0.021467  0.004577  0.015187  0.028119   \n",
       "R-4_H   0.034159  0.019174  0.014201  0.021362  0.004424  0.015591  0.028883   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "R4_tot  0.081727  0.051768  0.041783  0.057616  0.014798  0.038320  0.067987   \n",
       "R5_tot  0.081802  0.051783  0.041794  0.057669  0.014804  0.038295  0.067940   \n",
       "R6_tot  0.081889  0.051763  0.041819  0.057704  0.014796  0.038281  0.067898   \n",
       "R7_tot  0.081975  0.051712  0.041873  0.057743  0.014778  0.038265  0.067881   \n",
       "R8_tot  0.082043  0.051672  0.041885  0.057779  0.014747  0.038225  0.067933   \n",
       "\n",
       "               G         H         I         L         K         M         F  \\\n",
       "R-8_H   0.023597  0.008305  0.020903  0.035207  0.021388  0.008456  0.013840   \n",
       "R-7_H   0.023042  0.008143  0.020807  0.035373  0.021351  0.008483  0.013792   \n",
       "R-6_H   0.022268  0.008049  0.021052  0.036033  0.021327  0.008552  0.013814   \n",
       "R-5_H   0.021410  0.007920  0.021205  0.036702  0.021536  0.008618  0.013613   \n",
       "R-4_H   0.020069  0.007799  0.021404  0.037283  0.021816  0.008687  0.013470   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "R4_tot  0.071549  0.022834  0.060092  0.095151  0.059887  0.020857  0.039542   \n",
       "R5_tot  0.071696  0.022848  0.060033  0.095121  0.059897  0.020861  0.039552   \n",
       "R6_tot  0.071829  0.022907  0.059915  0.095113  0.059874  0.020858  0.039556   \n",
       "R7_tot  0.071950  0.022905  0.059925  0.095047  0.059776  0.020869  0.039515   \n",
       "R8_tot  0.072053  0.022922  0.059917  0.094951  0.059744  0.020856  0.039534   \n",
       "\n",
       "               P         S         T         W         Y         V  \n",
       "R-8_H   0.015724  0.022866  0.019575  0.004304  0.011312  0.023634  \n",
       "R-7_H   0.015211  0.022843  0.019518  0.004285  0.011227  0.023504  \n",
       "R-6_H   0.014686  0.022586  0.019409  0.004293  0.011017  0.023378  \n",
       "R-5_H   0.014068  0.022236  0.019130  0.004277  0.010902  0.023181  \n",
       "R-4_H   0.013333  0.021942  0.018862  0.004315  0.010845  0.023028  \n",
       "...          ...       ...       ...       ...       ...       ...  \n",
       "R4_tot  0.043656  0.061267  0.053969  0.013180  0.033308  0.070707  \n",
       "R5_tot  0.043544  0.061239  0.053944  0.013191  0.033342  0.070646  \n",
       "R6_tot  0.043509  0.061282  0.053904  0.013197  0.033330  0.070576  \n",
       "R7_tot  0.043529  0.061289  0.053884  0.013182  0.033347  0.070555  \n",
       "R8_tot  0.043510  0.061301  0.053866  0.013150  0.033359  0.070552  \n",
       "\n",
       "[68 rows x 20 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time \n",
    "aa=['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "secstr=[\"H\",\"E\",\"C\",\"tot\"]\n",
    "window=17\n",
    "gor_list_mat=[]\n",
    "for el in secstr:\n",
    "    gor_list_mat.append([el,np.zeros(((window),len(aa)),float)])   \n",
    "    \n",
    "def extracter(dssp,pssm):\n",
    "    for line in dssp:\n",
    "        if line.startswith(\">\"):\n",
    "            sec_str1=\"\"\n",
    "            sec_str=dssp.readline().rstrip()\n",
    "            for el in sec_str: \n",
    "                if el==\"-\":\n",
    "                    sec_str1=sec_str1+\"C\"\n",
    "                else:\n",
    "                    sec_str1=sec_str1+el \n",
    "    for line in pssm:\n",
    "        if \"A\" not in line:\n",
    "            line=line.rstrip()\n",
    "            line=line.split()[1:]\n",
    "            pssm_list.append(line)\n",
    "    return(sec_str1,pssm_list)\n",
    "\n",
    "def window_pssm(pssm_list,sec_str,window,el):\n",
    "        window_array=pssm_list[el-(window//2):el+window//2+1]\n",
    "        return(window_array,sec_str[el-window//2])\n",
    "\n",
    "def add_window_spaces(pssm_list,window):\n",
    "    for i in range(window//2):\n",
    "        pssm_list=[[0.0 for i in range(20)]]+pssm_list\n",
    "        pssm_list.append([0.0 for i in range(20)])\n",
    "    pssm_array=np.array([np.array(list) for list in pssm_list])\n",
    "    pssm_list_float=pssm_array.astype(np.float)\n",
    "    return(pssm_list_float)\n",
    "\n",
    "def fill_gor_mat(window_array,el_sec_str,gor_list_mat):\n",
    "    for list in range(len(gor_list_mat)):\n",
    "        if gor_list_mat[list][0]==el_sec_str:\n",
    "            gor_list_mat[list][1]+=window_array\n",
    "    gor_list_mat[3][1]+=window_array\n",
    "    return(gor_list_mat)\n",
    "\n",
    "def gor_df_fun(tot_gor_mat,secstr,aa,window):\n",
    "    index=[]\n",
    "    gor_df1=pd.DataFrame(tot_gor_mat)\n",
    "    gor_df1.columns=aa\n",
    "    for el in secstr:\n",
    "        for i in range(window):\n",
    "            index.append(\"R\"+str(i-(window//2))+\"_\"+el)\n",
    "    gor_df1.index=index\n",
    "    return(gor_df1)\n",
    "\n",
    "def normalization(gor_list_mat):\n",
    "    summation=[]\n",
    "    summation=np.sum(gor_list_mat[3][1],axis=1)\n",
    "    for array in gor_list_mat:\n",
    "        for sub in range(len(array[1])):\n",
    "            array[1][sub]=array[1][sub]/summation[sub]\n",
    "    return(gor_list_mat)\n",
    "\n",
    "prob_sec_str={}\n",
    "#print(prob_sec_str[1])\n",
    "for pssm_file in os.listdir(pssm_path):\n",
    "    #print(pssm_file)\n",
    "    pssm_list=[]\n",
    "    pssm_file=pssm_file.split(\".\")\n",
    "    pssm_file=\".\".join(pssm_file[:-1])\n",
    "    pssm=open(pssm_path+pssm_file+\".pssm\")\n",
    "    dssp=open(dssp_path+pssm_file+\".dssp\")\n",
    "    sec_str,pssm_list=extracter(dssp,pssm) # function recall\n",
    "    pssm_array=add_window_spaces(pssm_list,window) #function recall\n",
    "    for el in range(window//2,len(sec_str)+window//2):\n",
    "        window_array,el_sec_str=window_pssm(pssm_array,sec_str,window,el) #function recall\n",
    "        gor_list_mat=fill_gor_mat(window_array,el_sec_str,gor_list_mat) #function recall\n",
    "        keys=prob_sec_str.keys()\n",
    "        if el_sec_str in keys:\n",
    "            prob_sec_str[el_sec_str]+=1\n",
    "        else: \n",
    "            prob_sec_str[el_sec_str]=1\n",
    "#print(gor_list_mat)\n",
    "gor_list_mat=normalization(gor_list_mat) #function recall\n",
    "tot_gor_mat=np.concatenate((gor_list_mat[0][1],gor_list_mat[1][1],gor_list_mat[2][1],gor_list_mat[3][1]),axis=0)\n",
    "gor_df=gor_df_fun(tot_gor_mat,secstr,aa,window) #fucntion recall\n",
    "print(\"Number of the secondary structures: \",prob_sec_str)\n",
    "gor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to save and print the profile to a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "gor_df.to_csv(\"Gor_matrix.csv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prediction phase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to apply the formula to calculate the **Information function**:\n",
    "\n",
    "$I(S;R)= \\log\\frac{P(R,S)}{P(S)P(R)}$\n",
    "\n",
    "Where:\n",
    "- Each cell of the Gor matrix that we have generated is the $P(S;R)$\n",
    "- $P(S)$ (*secondary structure*) is the one that I have caluclated in the code cell above (*but can also be calculate by summing the row corresponding on the position 0 of each E,H and C submatrices*).\n",
    "- $P(R)$ (*probability of the residue*) is computed by summing the columns corresponding to each of the 20 residues that we have count at the position R_0."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will now calculate the **marginal probaility** $P(S)$ of the secondary structures (C,H, and E) from the count that I made in the training phase.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'C': 0.4240227763162528, 'H': 0.3547934387308039, 'E': 0.22118378495294336}\n"
     ]
    }
   ],
   "source": [
    "tot_sec_str=0\n",
    "for el in prob_sec_str:\n",
    "    tot_sec_str+=prob_sec_str[el]\n",
    "for el in prob_sec_str:\n",
    "    prob_sec_str[el]=prob_sec_str[el]/tot_sec_str\n",
    "print(prob_sec_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will now extract the $P(R_k)$ from the gor matrices (*gor_list_mat*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "info_list_fun=gor_list_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mat in info_list_fun:\n",
    "    if mat[0]==\"tot\":\n",
    "        tot_info_fun=mat[1]\n",
    "#print(tot_info_fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that I have both the **marginal probabilities** ($P(R)$ and $P(S)$) and the **join probability** $P(R,S)$ from the trained gor model I can compute the **information function** matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ar in (info_list_fun[:-1]):\n",
    "    for vec in range(len(ar[1])):\n",
    "        for el in range(len(ar[1][vec])):\n",
    "            P_ss=prob_sec_str[ar[0]]\n",
    "            P_join=ar[1][vec][el]\n",
    "            P_res=tot_info_fun[vec][el]\n",
    "            ar[1][vec][el]=math.log2(P_join/(P_ss*P_res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['H', array([[ 1.12899949e-01,  3.60889194e-02, -1.19113808e-02,\n",
      "         6.06961512e-02, -1.87352378e-01,  1.01648860e-01,\n",
      "         1.50661258e-01, -1.11056672e-01,  3.25065448e-02,\n",
      "        -2.46135331e-02,  7.39781451e-02,  2.31126941e-02,\n",
      "         1.33951036e-01, -1.29835795e-02,  1.29945522e-02,\n",
      "         5.70814166e-02,  1.52666965e-02, -1.09863415e-01,\n",
      "        -6.11998612e-02, -8.38485051e-02],\n",
      "       [ 1.44021032e-01,  4.16031656e-02, -1.61565469e-02,\n",
      "         6.68855721e-02, -1.89938072e-01,  1.18013262e-01,\n",
      "         1.79106789e-01, -1.43186399e-01,  4.71047247e-03,\n",
      "        -3.13506966e-02,  8.00712066e-02,  1.88369036e-02,\n",
      "         1.38697016e-01, -1.64531584e-02, -3.39645220e-02,\n",
      "         5.64504447e-02,  1.19516394e-02, -1.14363722e-01,\n",
      "        -7.19618367e-02, -9.03737781e-02],\n",
      "       [ 1.76082080e-01,  3.76835838e-02, -3.53023167e-02,\n",
      "         6.94593699e-02, -1.77307843e-01,  1.39754536e-01,\n",
      "         1.96575874e-01, -1.90597242e-01, -1.04675697e-02,\n",
      "        -1.58843040e-02,  1.05157345e-01,  1.57467755e-02,\n",
      "         1.50843910e-01, -1.65961754e-02, -8.24685118e-02,\n",
      "         4.13065049e-02,  6.23751751e-03, -1.11300080e-01,\n",
      "        -9.88419603e-02, -9.99258988e-02],\n",
      "       [ 2.09935453e-01,  5.63607041e-02, -5.03137956e-02,\n",
      "         6.36943291e-02, -1.92396848e-01,  1.64464561e-01,\n",
      "         2.27083499e-01, -2.44445040e-01, -3.52969358e-02,\n",
      "        -5.74465756e-03,  1.30487201e-01,  2.69121420e-02,\n",
      "         1.62253703e-01, -3.74674930e-02, -1.43074281e-01,\n",
      "         2.02881281e-02, -1.35411875e-02, -1.18884736e-01,\n",
      "        -1.15917553e-01, -1.13008466e-01],\n",
      "       [ 2.38990196e-01,  7.21586238e-02, -6.36458547e-02,\n",
      "         5.79548360e-02, -2.42430486e-01,  2.01529572e-01,\n",
      "         2.65267449e-01, -3.35019153e-01, -5.62509624e-02,\n",
      "         6.16538079e-03,  1.51517676e-01,  4.40944566e-02,\n",
      "         1.75278503e-01, -5.36635297e-02, -2.20564494e-01,\n",
      "         1.23113540e-03, -3.38209968e-02, -1.04746140e-01,\n",
      "        -1.22606400e-01, -1.22923579e-01],\n",
      "       [ 2.95286247e-01,  8.01509038e-02, -1.10492076e-01,\n",
      "         3.33989077e-02, -2.52771533e-01,  2.15777924e-01,\n",
      "         2.81796985e-01, -4.55875363e-01, -7.97602517e-02,\n",
      "         3.88612817e-02,  2.01429936e-01,  4.35916886e-02,\n",
      "         2.01491599e-01, -3.32233121e-02, -3.43498189e-01,\n",
      "        -3.91626244e-02, -7.18195427e-02, -6.82153758e-02,\n",
      "        -1.15597388e-01, -1.25805192e-01],\n",
      "       [ 3.48205266e-01,  9.86045225e-02, -1.52207522e-01,\n",
      "         1.04089708e-03, -2.62099867e-01,  2.31486396e-01,\n",
      "         2.98376640e-01, -6.31544594e-01, -1.38476227e-01,\n",
      "         6.15289480e-02,  2.77185286e-01,  5.33692838e-02,\n",
      "         2.37535503e-01, -1.92196710e-02, -5.07378140e-01,\n",
      "        -7.73382703e-02, -1.26526856e-01, -1.32340087e-02,\n",
      "        -1.15266326e-01, -1.43647869e-01],\n",
      "       [ 3.85267956e-01,  1.53617642e-01, -1.83918264e-01,\n",
      "        -3.47711104e-02, -2.96294215e-01,  2.67553227e-01,\n",
      "         3.37550692e-01, -7.78929644e-01, -1.48513902e-01,\n",
      "         6.42499481e-02,  2.91622236e-01,  1.13004924e-01,\n",
      "         2.00065286e-01, -3.40420590e-02, -7.37592080e-01,\n",
      "        -1.02121093e-01, -1.65337613e-01,  1.89996575e-03,\n",
      "        -1.21112570e-01, -1.65488238e-01],\n",
      "       [ 4.43428579e-01,  2.04680947e-01, -3.51321346e-01,\n",
      "        -2.70755697e-01, -3.07608643e-01,  3.17784929e-01,\n",
      "         3.72736251e-01, -9.66749409e-01, -1.47744123e-01,\n",
      "         1.12102453e-01,  3.79630225e-01,  1.75618804e-01,\n",
      "         2.72309334e-01,  2.96378694e-02, -1.03213169e+00,\n",
      "        -2.79112257e-01, -3.06517551e-01,  2.92779701e-02,\n",
      "        -6.73019773e-02, -1.26773221e-01],\n",
      "       [ 4.49047429e-01,  2.17075130e-01, -2.03639633e-01,\n",
      "        -2.69153453e-01, -2.17396958e-01,  3.47414766e-01,\n",
      "         3.50686695e-01, -6.91415282e-01, -7.39654213e-02,\n",
      "         8.19489223e-02,  3.79379575e-01,  1.86747225e-01,\n",
      "         3.37365824e-01,  1.66252928e-02, -1.74151293e+00,\n",
      "        -2.27902740e-01, -3.16719870e-01, -3.42514253e-02,\n",
      "        -6.77979329e-02, -1.62425359e-01],\n",
      "       [ 4.20969218e-01,  2.27045529e-01, -1.44961335e-01,\n",
      "        -3.38479614e-01, -1.83726688e-01,  3.29932777e-01,\n",
      "         2.10229845e-01, -4.32795374e-01, -5.31965378e-02,\n",
      "         9.59792794e-02,  3.90672200e-01,  2.02125263e-01,\n",
      "         3.56028528e-01,  2.30664603e-02, -1.50767717e+00,\n",
      "        -2.31356505e-01, -3.12650499e-01, -6.49132008e-02,\n",
      "        -8.05080362e-02, -1.60687644e-01],\n",
      "       [ 3.90718637e-01,  2.35792303e-01, -1.04648377e-01,\n",
      "        -3.90457260e-01, -1.58824027e-01,  2.67219911e-01,\n",
      "         1.13037130e-01, -3.48539580e-01, -4.14956806e-02,\n",
      "         1.16221591e-01,  3.97097509e-01,  2.29626970e-01,\n",
      "         3.58014143e-01,  4.27640756e-02, -1.20278179e+00,\n",
      "        -2.10440547e-01, -3.27403722e-01, -7.07204459e-02,\n",
      "        -6.94007337e-02, -1.61088814e-01],\n",
      "       [ 3.27250978e-01,  2.06748858e-01, -4.11679309e-02,\n",
      "        -2.68831373e-01, -1.91465739e-01,  2.60495995e-01,\n",
      "         1.63794267e-01, -2.61465779e-01, -3.09510788e-03,\n",
      "         4.00610203e-02,  3.21262508e-01,  2.52646867e-01,\n",
      "         2.95718043e-01, -1.10379888e-02, -7.97701734e-01,\n",
      "        -1.50774110e-01, -2.89866465e-01, -1.40192617e-01,\n",
      "        -9.84525086e-02, -2.19818132e-01],\n",
      "       [ 2.77777777e-01,  1.60290303e-01, -3.40742138e-02,\n",
      "        -2.30013216e-01, -1.83915490e-01,  2.13721365e-01,\n",
      "         1.35381849e-01, -2.04072951e-01, -9.25174521e-03,\n",
      "         4.09771476e-02,  2.88081878e-01,  2.24823222e-01,\n",
      "         2.43409998e-01,  2.40074610e-03, -5.45893243e-01,\n",
      "        -1.25568994e-01, -2.52150962e-01, -1.52904171e-01,\n",
      "        -7.13549785e-02, -1.88455336e-01],\n",
      "       [ 2.21678104e-01,  1.02400345e-01, -4.14666742e-02,\n",
      "        -2.15876531e-01, -1.41339228e-01,  1.64048592e-01,\n",
      "         9.62600915e-02, -1.75865724e-01, -4.58736723e-03,\n",
      "         8.31344171e-02,  2.86066165e-01,  1.71231757e-01,\n",
      "         2.23667029e-01,  2.43318766e-02, -4.14548979e-01,\n",
      "        -1.12065165e-01, -2.30001771e-01, -1.24065026e-01,\n",
      "        -6.02556124e-02, -1.24068598e-01],\n",
      "       [ 1.68469695e-01,  9.30822864e-02, -2.28020603e-02,\n",
      "        -1.55811066e-01, -1.33571295e-01,  1.75844765e-01,\n",
      "         1.23892271e-01, -1.45413928e-01, -5.16970317e-03,\n",
      "         5.21964171e-02,  2.11717852e-01,  1.85242636e-01,\n",
      "         1.49069193e-01, -1.39771039e-02, -2.91366883e-01,\n",
      "        -8.07627765e-02, -1.90645596e-01, -1.78674797e-01,\n",
      "        -8.37743079e-02, -1.00869420e-01],\n",
      "       [ 1.33612618e-01,  8.03592289e-02, -7.65874640e-03,\n",
      "        -1.02186695e-01, -1.57506987e-01,  1.78693458e-01,\n",
      "         1.38079307e-01, -1.35064294e-01,  5.85031972e-03,\n",
      "         2.83048118e-02,  1.51187904e-01,  1.79793093e-01,\n",
      "         9.53993268e-02, -3.69234242e-02, -2.19076901e-01,\n",
      "        -6.01256906e-02, -1.53501428e-01, -1.61188777e-01,\n",
      "        -9.65705770e-02, -8.27833477e-02]])], ['E', array([[-1.20551421e-01, -3.73810342e-02,  1.04141269e-01,\n",
      "         7.54051553e-02,  7.39014108e-02, -4.53754411e-02,\n",
      "        -4.68996624e-02,  1.80899477e-01, -2.39898634e-02,\n",
      "        -1.17200512e-01, -1.62660331e-01,  3.54386633e-02,\n",
      "        -1.63472726e-01, -3.39982576e-02,  1.23447848e-01,\n",
      "         8.42150386e-02,  5.75466654e-02,  3.92096103e-02,\n",
      "         3.98707492e-02, -2.25458795e-02],\n",
      "       [-1.56726196e-01, -4.51526814e-02,  2.07935703e-01,\n",
      "         1.26961885e-01,  6.40401059e-02, -4.42214598e-02,\n",
      "        -5.89634465e-02,  3.00908268e-01,  3.16237546e-02,\n",
      "        -2.09316043e-01, -2.30739328e-01,  7.23777073e-02,\n",
      "        -2.09583396e-01, -9.37293233e-02,  1.55068314e-01,\n",
      "         1.05615528e-01,  6.22602739e-02, -1.08767302e-02,\n",
      "        -9.83673984e-03, -7.58057787e-02],\n",
      "       [-2.23063695e-01, -4.51773084e-02,  2.64076160e-01,\n",
      "         1.79799130e-01,  4.88773067e-02, -4.12711942e-02,\n",
      "        -6.27735823e-02,  4.22733135e-01,  6.13297981e-02,\n",
      "        -2.52750293e-01, -3.20767463e-01,  1.11799300e-01,\n",
      "        -2.14141660e-01, -1.64302962e-01,  2.02076688e-01,\n",
      "         1.16919916e-01,  7.72111769e-02, -1.22775079e-01,\n",
      "        -6.44988351e-02, -1.02239012e-01],\n",
      "       [-2.61148532e-01, -4.69492728e-03,  2.68877772e-01,\n",
      "         1.76737441e-01, -3.71657936e-02, -6.07079212e-02,\n",
      "        -7.72914882e-02,  4.91051548e-01,  7.76052550e-02,\n",
      "        -2.42536160e-01, -3.96022630e-01,  1.34978923e-01,\n",
      "        -2.46820304e-01, -1.99007139e-01,  2.41611050e-01,\n",
      "         1.12757659e-01,  1.07787873e-01, -1.84565937e-01,\n",
      "        -7.08764531e-02, -6.62620623e-02],\n",
      "       [-2.72704117e-01, -2.06585795e-02,  1.76475406e-01,\n",
      "         7.89973415e-02, -2.35886947e-02, -9.60213658e-02,\n",
      "        -1.50223400e-01,  5.16269214e-01,  9.23666721e-02,\n",
      "        -1.33912044e-01, -3.67884196e-01,  1.05966267e-01,\n",
      "        -2.26653615e-01, -1.55675621e-01,  2.09453967e-01,\n",
      "         6.99734257e-02,  1.30894271e-01, -1.61043193e-01,\n",
      "         2.34396433e-02,  8.02296065e-02],\n",
      "       [-2.99988489e-01, -3.07725438e-02,  1.07848877e-01,\n",
      "        -2.34731262e-02,  2.12434077e-03, -1.49598561e-01,\n",
      "        -2.29839771e-01,  5.04542417e-01,  6.74313441e-02,\n",
      "         4.99047047e-02, -3.20354423e-01,  4.23499343e-02,\n",
      "        -2.03186523e-01, -7.40501112e-02,  2.61997944e-02,\n",
      "         1.23222097e-03,  1.56041871e-01, -2.70800851e-02,\n",
      "         1.28571446e-01,  2.68270688e-01],\n",
      "       [-2.96656458e-01, -7.93306664e-02, -1.08414376e-01,\n",
      "        -2.68484494e-01,  7.68798232e-02, -2.55944467e-01,\n",
      "        -3.81320661e-01,  4.05598780e-01,  9.07008217e-04,\n",
      "         3.24396754e-01, -1.59176209e-01, -9.50320260e-02,\n",
      "        -1.26037500e-01,  1.30688158e-01, -1.71100496e-01,\n",
      "        -1.44088040e-01,  1.32946355e-01,  1.53975023e-01,\n",
      "         2.80157698e-01,  5.26821348e-01],\n",
      "       [-3.18633710e-01, -6.68133787e-02, -4.40355634e-01,\n",
      "        -6.10889370e-01,  2.16770938e-01, -2.91392596e-01,\n",
      "        -4.44218082e-01, -4.20540068e-02, -5.38367908e-03,\n",
      "         5.50526563e-01, -3.54593960e-03, -1.72033738e-01,\n",
      "        -7.32772441e-04,  3.21132660e-01, -4.47703268e-01,\n",
      "        -2.03386292e-01,  2.04339710e-01,  3.46289813e-01,\n",
      "         4.37586443e-01,  7.40188786e-01],\n",
      "       [-3.48282495e-01, -1.06347398e-01, -5.54480644e-01,\n",
      "        -7.15742632e-01,  3.51322580e-01, -3.65605717e-01,\n",
      "        -5.11201413e-01, -4.35599357e-01, -2.04975280e-02,\n",
      "         6.76013015e-01,  9.14990166e-02, -2.99930090e-01,\n",
      "        -2.78508619e-02,  4.41473711e-01, -7.42839074e-01,\n",
      "        -1.71945907e-01,  2.36823897e-01,  3.89548965e-01,\n",
      "         5.23800671e-01,  8.43960488e-01],\n",
      "       [-3.05444388e-01, -1.99537430e-01, -2.76555207e-01,\n",
      "        -3.14614835e-01,  4.23263402e-01, -4.37797726e-01,\n",
      "        -5.10296526e-01, -1.78683968e-01, -3.13689156e-02,\n",
      "         5.62930366e-01,  6.56322023e-02, -3.71448530e-01,\n",
      "        -5.19074100e-02,  3.22544869e-01, -4.74477883e-01,\n",
      "        -5.60794024e-02,  2.33594137e-01,  2.52141408e-01,\n",
      "         3.98075629e-01,  7.27527805e-01],\n",
      "       [-2.93670466e-01, -2.11713343e-01, -2.96853165e-02,\n",
      "         3.63823771e-02,  2.83209636e-01, -4.24445681e-01,\n",
      "        -3.58549909e-01,  1.72180047e-01,  4.64952713e-02,\n",
      "         2.80518693e-01, -1.49055943e-01, -2.98873338e-01,\n",
      "        -1.89999172e-01,  1.50699208e-01,  5.14684774e-02,\n",
      "         1.21742517e-01,  2.29958822e-01,  3.50136020e-02,\n",
      "         2.11259410e-01,  4.56026987e-01],\n",
      "       [-2.92123392e-01, -1.93976554e-01,  1.71232116e-01,\n",
      "         2.83070298e-01,  2.02789982e-01, -3.14320398e-01,\n",
      "        -2.00151280e-01,  3.42561564e-01,  7.48633929e-02,\n",
      "         9.50441012e-03, -3.35139802e-01, -1.97047343e-01,\n",
      "        -2.93584166e-01,  1.34903196e-02,  1.97196635e-01,\n",
      "         2.05499891e-01,  2.38347746e-01, -9.38074171e-02,\n",
      "         1.55822837e-02,  2.18120934e-01],\n",
      "       [-2.83438931e-01, -1.76313501e-01,  2.51987182e-01,\n",
      "         3.21705335e-01,  1.28766696e-01, -2.40607367e-01,\n",
      "        -1.54204540e-01,  5.04052949e-01,  9.32708585e-02,\n",
      "        -1.84219348e-01, -4.24843607e-01, -1.39304430e-01,\n",
      "        -3.17510605e-01, -8.00747936e-02,  2.85171056e-01,\n",
      "         2.27698933e-01,  2.35687324e-01, -9.91143568e-02,\n",
      "        -3.04538953e-02,  2.44541844e-02],\n",
      "       [-3.00128696e-01, -1.48062949e-01,  2.64962952e-01,\n",
      "         3.30725659e-01,  9.73864481e-02, -1.72410012e-01,\n",
      "        -8.37581614e-02,  4.85437182e-01,  7.92855702e-02,\n",
      "        -2.57721446e-01, -4.35214653e-01, -6.91702098e-02,\n",
      "        -2.79943109e-01, -1.36679316e-01,  3.09950229e-01,\n",
      "         2.34051180e-01,  2.34172660e-01, -1.12540866e-01,\n",
      "        -3.40843966e-02, -5.92879102e-02],\n",
      "       [-2.69645799e-01, -1.07371130e-01,  2.17474498e-01,\n",
      "         2.62686125e-01,  4.53859349e-02, -1.42977752e-01,\n",
      "        -4.87637258e-02,  4.25673501e-01,  9.28798277e-02,\n",
      "        -2.47071070e-01, -3.64575780e-01, -3.75685120e-02,\n",
      "        -2.60070438e-01, -1.67289717e-01,  3.25487602e-01,\n",
      "         1.99706050e-01,  2.08511813e-01, -3.56288812e-02,\n",
      "        -6.77184897e-03, -7.01882115e-02],\n",
      "       [-2.41690257e-01, -7.63868586e-02,  1.58310139e-01,\n",
      "         1.99116563e-01, -3.49079471e-02, -1.23232437e-01,\n",
      "        -3.92487158e-02,  3.55450227e-01,  1.00227061e-01,\n",
      "        -1.90327807e-01, -3.07309690e-01, -3.14074548e-02,\n",
      "        -2.32377545e-01, -1.23313948e-01,  3.04832514e-01,\n",
      "         1.67238308e-01,  1.86894774e-01,  6.23935165e-02,\n",
      "         2.97362762e-02, -3.79408873e-02],\n",
      "       [-1.88244374e-01, -5.97044541e-02,  1.03266202e-01,\n",
      "         1.53239195e-01, -2.49513913e-02, -1.06469272e-01,\n",
      "        -3.92321632e-02,  2.97713493e-01,  3.82710129e-02,\n",
      "        -1.49666102e-01, -2.47384124e-01, -5.50682089e-02,\n",
      "        -1.94841704e-01, -8.64432910e-02,  2.47019519e-01,\n",
      "         1.34671899e-01,  1.64592253e-01,  1.46959578e-01,\n",
      "         8.41887299e-02,  6.22743049e-03]])], ['C', array([[-0.03844142, -0.01137363, -0.04716928, -0.09533111,  0.10364408,\n",
      "        -0.06632118, -0.11313179, -0.01113902, -0.01517715,  0.07704412,\n",
      "         0.01662975, -0.03872483, -0.03730863,  0.02806748, -0.08034818,\n",
      "        -0.09715515, -0.04413512,  0.06623579,  0.02876045,  0.07769778],\n",
      "       [-0.05011849, -0.01218044, -0.107108  , -0.13240599,  0.11041674,\n",
      "        -0.0824934 , -0.13560506, -0.0617816 , -0.02077479,  0.12225451,\n",
      "         0.04172762, -0.05564171, -0.02019924,  0.05977106, -0.05847778,\n",
      "        -0.10937229, -0.04389069,  0.09448679,  0.06247688,  0.10769881],\n",
      "       [-0.04974154, -0.00877232, -0.12745082, -0.16894104,  0.10941593,\n",
      "        -0.10531726, -0.15189404, -0.11098704, -0.02415555,  0.12828142,\n",
      "         0.05758723, -0.07584436, -0.02954248,  0.09182703, -0.04683214,\n",
      "        -0.10212121, -0.04737281,  0.14362792,  0.10864302,  0.12655032],\n",
      "       [-0.06596317, -0.04638478, -0.11745695, -0.16125074,  0.16057413,\n",
      "        -0.11957436, -0.17694197, -0.12200896, -0.01246795,  0.11642259,\n",
      "         0.06489614, -0.09991946, -0.02557903,  0.12255732, -0.023419  ,\n",
      "        -0.08047667, -0.047888  ,  0.17482429,  0.12380792,  0.11963179],\n",
      "       [-0.09066235, -0.05214458, -0.04656887, -0.09488971,  0.18635509,\n",
      "        -0.13900096, -0.1800344 , -0.07558099, -0.0036005 ,  0.06026278,\n",
      "         0.03527538, -0.09808055, -0.04729895,  0.11624324,  0.05258093,\n",
      "        -0.03895173, -0.04417125,  0.15535195,  0.08355659,  0.0545152 ],\n",
      "       [-0.13941675, -0.05408192,  0.03026731, -0.01621748,  0.18105011,\n",
      "        -0.12596617, -0.15789654,  0.01193127,  0.02863136, -0.06070518,\n",
      "        -0.03136048, -0.06070932, -0.08449911,  0.0637062 ,  0.22369509,\n",
      "         0.0313419 , -0.02760132,  0.06811031,  0.02262587, -0.05398626],\n",
      "       [-0.20332875, -0.04584702,  0.16548415,  0.12164378,  0.1511474 ,\n",
      "        -0.0905433 , -0.10638982,  0.17286159,  0.10601483, -0.26531339,\n",
      "        -0.18894319,  0.00248015, -0.16184457, -0.05647   ,  0.38562759,\n",
      "         0.12862854,  0.02843258, -0.07568952, -0.07028325, -0.23550207],\n",
      "       [-0.23914594, -0.10532509,  0.30726894,  0.2630301 ,  0.09865467,\n",
      "        -0.11262541, -0.12395104,  0.44791907,  0.11608876, -0.47486619,\n",
      "        -0.29718417, -0.01382836, -0.19132548, -0.1689919 ,  0.55981225,\n",
      "         0.17091209,  0.01611197, -0.22198429, -0.18026351, -0.43249287],\n",
      "       [-0.30258699, -0.13681566,  0.43006897,  0.42995524,  0.02409269,\n",
      "        -0.1351307 , -0.13770918,  0.6271922 ,  0.12274647, -0.6882612 ,\n",
      "        -0.4860405 , -0.0150816 , -0.25816445, -0.32935272,  0.71149949,\n",
      "         0.26997989,  0.0936968 , -0.28428492, -0.30436386, -0.60887964],\n",
      "       [-0.33277476, -0.10234868,  0.26440833,  0.31569619, -0.09050914,\n",
      "        -0.13811255, -0.11221876,  0.46399344,  0.07455585, -0.51017339,\n",
      "        -0.46584427,  0.00399862, -0.32554047, -0.21823721,  0.79302321,\n",
      "         0.19175808,  0.10197713, -0.12027349, -0.19650231, -0.42080436],\n",
      "       [-0.29944484, -0.10688939,  0.12513811,  0.2161881 , -0.01900949,\n",
      "        -0.1235147 , -0.02396514,  0.2025595 ,  0.01892744, -0.26797913,\n",
      "        -0.33829404, -0.04131839, -0.26852697, -0.10612694,  0.60741541,\n",
      "         0.10841985,  0.10150101,  0.03421456, -0.05416233, -0.16121528],\n",
      "       [-0.25954183, -0.12491112, -0.01041204,  0.11811939,  0.01221038,\n",
      "        -0.1018205 , -0.00089947,  0.05627836, -0.00586782, -0.1103603 ,\n",
      "        -0.24638354, -0.11679871, -0.21696693, -0.04405346,  0.48214844,\n",
      "         0.04783494,  0.10591374,  0.10146962,  0.0477248 ,  0.00469755],\n",
      "       [-0.18443392, -0.10302243, -0.11405464,  0.01712972,  0.07764489,\n",
      "        -0.12887055, -0.07054351, -0.119425  , -0.04847868,  0.05515454,\n",
      "        -0.11347711, -0.17158005, -0.13179741,  0.04899082,  0.3119499 ,\n",
      "        -0.00889965,  0.08369823,  0.15330769,  0.09233103,  0.14969518],\n",
      "       [-0.11980364, -0.07012839, -0.12915465, -0.01608101,  0.08911917,\n",
      "        -0.11223872, -0.07840937, -0.1492645 , -0.03522553,  0.08573501,\n",
      "        -0.07261864, -0.17899498, -0.09199753,  0.06454703,  0.18723718,\n",
      "        -0.0322834 ,  0.05969733,  0.16772775,  0.07389832,  0.16795427],\n",
      "       [-0.07407041, -0.03524799, -0.09104911,  0.017368  ,  0.08598927,\n",
      "        -0.07635341, -0.05948994, -0.12500189, -0.04696661,  0.04609933,\n",
      "        -0.10008613, -0.13929347, -0.08050509,  0.06057277,  0.10695364,\n",
      "        -0.02167047,  0.05962455,  0.11326367,  0.05195666,  0.12924372],\n",
      "       [-0.03383938, -0.04225591, -0.07002378,  0.01215283,  0.11963579,\n",
      "        -0.09812718, -0.09084997, -0.09774899, -0.05070319,  0.04773824,\n",
      "        -0.04719336, -0.15747314, -0.01957802,  0.07148983,  0.04304799,\n",
      "        -0.02702572,  0.04464406,  0.10354867,  0.05149779,  0.09766899],\n",
      "       [-0.02530131, -0.03916577, -0.05031766, -0.00179012,  0.13156934,\n",
      "        -0.10970189, -0.10476977, -0.06608887, -0.02535716,  0.04938515,\n",
      "        -0.01486044, -0.13846793,  0.01249501,  0.07242172,  0.02922923,\n",
      "        -0.02456518,  0.03057068,  0.04616453,  0.03256613,  0.06268061]])], ['tot', array([[0.08167044, 0.05122172, 0.04194934, 0.05803638, 0.01468446,\n",
      "        0.0381463 , 0.06754626, 0.07182992, 0.02288732, 0.05993113,\n",
      "        0.09427139, 0.0593258 , 0.02172149, 0.0393621 , 0.04392247,\n",
      "        0.06194793, 0.05459309, 0.01308938, 0.03326389, 0.07059919],\n",
      "       [0.08165844, 0.05128744, 0.04194214, 0.05805958, 0.01470556,\n",
      "        0.03818415, 0.06764637, 0.07171982, 0.02287766, 0.0599344 ,\n",
      "        0.09431643, 0.05939893, 0.02171861, 0.03931842, 0.04389428,\n",
      "        0.06191392, 0.05455909, 0.01307287, 0.03326275, 0.07052913],\n",
      "       [0.08160537, 0.05137348, 0.04189416, 0.0579585 , 0.01473952,\n",
      "        0.03816731, 0.06771308, 0.07162837, 0.02285127, 0.05999216,\n",
      "        0.09442014, 0.05945855, 0.02171156, 0.0393865 , 0.04382704,\n",
      "        0.06186232, 0.0544696 , 0.01307159, 0.03325323, 0.07061627],\n",
      "       [0.08162449, 0.05141685, 0.04183255, 0.05789142, 0.01473974,\n",
      "        0.03819351, 0.06771264, 0.0714883 , 0.02287529, 0.06000566,\n",
      "        0.09449973, 0.05957772, 0.02170696, 0.03937786, 0.04378493,\n",
      "        0.06179879, 0.05442611, 0.01308994, 0.03329857, 0.07065893],\n",
      "       [0.08158115, 0.05140699, 0.04183265, 0.05783787, 0.0147502 ,\n",
      "        0.03821361, 0.06773531, 0.07134986, 0.02285453, 0.06007081,\n",
      "        0.09460659, 0.05963797, 0.02168236, 0.03940353, 0.04378592,\n",
      "        0.06179281, 0.05442379, 0.01307738, 0.03327956, 0.07067709],\n",
      "       [0.08154421, 0.05144353, 0.04177024, 0.05778031, 0.01476973,\n",
      "        0.03822495, 0.06780759, 0.07124561, 0.02284709, 0.06008931,\n",
      "        0.09468017, 0.05976091, 0.02168781, 0.03942834, 0.04379319,\n",
      "        0.06170642, 0.05439917, 0.01307976, 0.03328721, 0.07065445],\n",
      "       [0.08158042, 0.05150094, 0.04170917, 0.057669  , 0.01477894,\n",
      "        0.03823667, 0.06786529, 0.07110627, 0.02283393, 0.06012138,\n",
      "        0.09470971, 0.05989255, 0.02168864, 0.03940528, 0.04382983,\n",
      "        0.06170761, 0.05431682, 0.01309705, 0.03328661, 0.07066389],\n",
      "       [0.08156181, 0.0515497 , 0.04171807, 0.05759939, 0.01480236,\n",
      "        0.03826252, 0.06786965, 0.07103778, 0.02282768, 0.06006214,\n",
      "        0.09474191, 0.06002714, 0.02167407, 0.03941074, 0.04388851,\n",
      "        0.0617081 , 0.05427852, 0.01310257, 0.03324083, 0.0706365 ],\n",
      "       [0.08153256, 0.05164816, 0.04170454, 0.05753416, 0.01479424,\n",
      "        0.03829976, 0.06792565, 0.0710566 , 0.02280955, 0.05999359,\n",
      "        0.09473831, 0.06015655, 0.02166397, 0.03941604, 0.04397558,\n",
      "        0.0616885 , 0.05419629, 0.01310775, 0.03319593, 0.07056229],\n",
      "       [0.08149751, 0.05175535, 0.04172117, 0.05751759, 0.01482988,\n",
      "        0.03834973, 0.0679486 , 0.07108754, 0.02282157, 0.06010836,\n",
      "        0.09496239, 0.06022174, 0.02104419, 0.0394948 , 0.04391378,\n",
      "        0.06148783, 0.05417939, 0.01313798, 0.03326822, 0.07065239],\n",
      "       [0.08156629, 0.05174502, 0.04168366, 0.05750273, 0.01482533,\n",
      "        0.03837147, 0.06793682, 0.07128961, 0.02281798, 0.06014557,\n",
      "        0.09511246, 0.06007159, 0.02092925, 0.03952967, 0.04382899,\n",
      "        0.06141711, 0.05407029, 0.01316251, 0.03332217, 0.07067148],\n",
      "       [0.08164528, 0.05171565, 0.0417337 , 0.05755628, 0.01480605,\n",
      "        0.03834143, 0.06799503, 0.07143301, 0.0228297 , 0.060152  ,\n",
      "        0.09515928, 0.05992759, 0.02089038, 0.03955929, 0.04369756,\n",
      "        0.06132461, 0.05400167, 0.01317664, 0.03334773, 0.07070713],\n",
      "       [0.08172726, 0.05176781, 0.04178347, 0.0576162 , 0.01479796,\n",
      "        0.0383196 , 0.06798708, 0.07154928, 0.0228339 , 0.06009233,\n",
      "        0.09515094, 0.05988737, 0.02085679, 0.03954248, 0.04365635,\n",
      "        0.06126739, 0.05396925, 0.01317968, 0.03330812, 0.07070674],\n",
      "       [0.08180176, 0.05178271, 0.0417936 , 0.05766892, 0.01480354,\n",
      "        0.03829495, 0.0679403 , 0.0716964 , 0.0228481 , 0.06003326,\n",
      "        0.09512134, 0.05989685, 0.02086107, 0.03955197, 0.04354435,\n",
      "        0.06123861, 0.05394359, 0.01319083, 0.03334217, 0.07064568],\n",
      "       [0.08188877, 0.05176306, 0.04181898, 0.05770447, 0.01479607,\n",
      "        0.03828104, 0.06789812, 0.07182873, 0.02290697, 0.05991511,\n",
      "        0.09511348, 0.05987433, 0.02085761, 0.03955572, 0.04350874,\n",
      "        0.06128172, 0.05390413, 0.0131967 , 0.03333031, 0.07057592],\n",
      "       [0.08197456, 0.05171178, 0.04187326, 0.05774345, 0.01477771,\n",
      "        0.03826479, 0.06788102, 0.07195042, 0.02290518, 0.05992531,\n",
      "        0.09504722, 0.05977567, 0.02086891, 0.03951508, 0.04352882,\n",
      "        0.0612885 , 0.05388423, 0.0131819 , 0.03334691, 0.07055527],\n",
      "       [0.08204301, 0.05167193, 0.0418848 , 0.05777931, 0.01474711,\n",
      "        0.03822504, 0.06793328, 0.0720532 , 0.02292164, 0.05991745,\n",
      "        0.09495131, 0.05974387, 0.02085561, 0.03953399, 0.04351045,\n",
      "        0.06130127, 0.05386586, 0.0131497 , 0.0333591 , 0.07055207]])]]\n"
     ]
    }
   ],
   "source": [
    "print(info_list_fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To save the information function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that I have calculated the information function for each cell of the gor model, I can perform the prediction of the secondary structure of a new sequence (*from the blind set*) from the profile that I have generated using *psiblast*. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predicted conformation of a residue R at position j is the one having the **highest value of the window-based information function**:\n",
    "\n",
    "$S* = argmax_s I(S;R_{-d},...,R_d)= argmax_s\\sum_{k=-d}^d I(S;R_k)$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the prediction phase indeed also the profile of the sequences (*blind-set*) for which we want to predict the secondary structure is taken into consideration:\n",
    "\n",
    "\n",
    "$PW$ (Profile Window) $= \\begin{matrix} -1 & A:0.2 & I:0.8 \\\\ 0 & D:0.9 & N:0.1 \\\\ +1 & E:0.3 & P:0.7 \\end{matrix}$  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$I(H;PW)= \\sum_{k=-1}^1 \\sum_R PW_{k,R} \\times I(H;R_k) = 0.2 \\times I(H;A_{-1}) + 0.8 \\times I(H;I_{-1}) + 0.9 \\times I(H;D_0) + 0.1 \\times I(H;N_0) + 0.3 \\times I(H;E_1) + 0.7 \\times I(H;P_1)$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prediction phase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "prof_path=\"Blindset_pssm/\" #Path of the sequence profile for which we wanna predict the secondary structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predicted Secondary Structure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4XO2_A</th>\n",
       "      <td>HHHHHHHHHHHHHHHHHHHHHHHHHHHCCCCEEEEHHHHHHHHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4XP8_A</th>\n",
       "      <td>ECHHHHHCCEEEHCHHHHHHHHHHHHCCCEEEEEEEEECCCEEEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4XPU_A</th>\n",
       "      <td>HHHHHHHHHHHHHHHHHHHHCCCCCCCEEEEEEEEEEEECCCHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4XPW_A</th>\n",
       "      <td>CCEEEECCCCHEEEEEHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4Y7M_C</th>\n",
       "      <td>HHHHHHHHHHHEEEEEEEEEEECCCCEEEEEECCCCHHHHHHHEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7BQI_A</th>\n",
       "      <td>HHHHHHHHHHHHHHHHHHHHHHHHCCCEEEEECCCHHHHHHHHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5D1V_A</th>\n",
       "      <td>HHHHHHHHCCCHHHHHHHHHHHHHEHHHHHHEEEEHHHHHHHHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5K31_A</th>\n",
       "      <td>EHCCCHHHHHHHHHHHHHHHEEECCCCCCCCCHHEHHHHEEECCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5V89_A</th>\n",
       "      <td>HHHHHHHHHHHHHCCCCCEEEHCCHEEEEEHCCCCCHHHHHHHHHH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6H8O_A</th>\n",
       "      <td>HHEEEEEEEECCCCCCEEEHHHHHHEEEEEECCCEECCCHHHHHHH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Predicted Secondary Structure\n",
       "4XO2_A  HHHHHHHHHHHHHHHHHHHHHHHHHHHCCCCEEEEHHHHHHHHHHH...\n",
       "4XP8_A  ECHHHHHCCEEEHCHHHHHHHHHHHHCCCEEEEEEEEECCCEEEEE...\n",
       "4XPU_A  HHHHHHHHHHHHHHHHHHHHCCCCCCCEEEEEEEEEEEECCCHHHH...\n",
       "4XPW_A  CCEEEECCCCHEEEEEHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH...\n",
       "4Y7M_C  HHHHHHHHHHHEEEEEEEEEEECCCCEEEEEECCCCHHHHHHHEEE...\n",
       "...                                                   ...\n",
       "7BQI_A  HHHHHHHHHHHHHHHHHHHHHHHHCCCEEEEECCCHHHHHHHHHHH...\n",
       "5D1V_A  HHHHHHHHCCCHHHHHHHHHHHHHEHHHHHHEEEEHHHHHHHHHHH...\n",
       "5K31_A  EHCCCHHHHHHHHHHHHHHHEEECCCCCCCCCHHEHHHHEEECCCC...\n",
       "5V89_A  HHHHHHHHHHHHHCCCCCEEEHCCHEEEEEHCCCCCHHHHHHHHHH...\n",
       "6H8O_A  HHEEEEEEEECCCCCCEEEHHHHHHEEEEEECCCEECCCHHHHHHH...\n",
       "\n",
       "[126 rows x 1 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import math\n",
    "#window=17\n",
    "def reading_profile(profile):\n",
    "    prof_list=[]\n",
    "    long=0\n",
    "    for line in profile:\n",
    "        if \"A\" not in line:\n",
    "            #prof_line=[]\n",
    "            line=line.rstrip()\n",
    "            array=line.split()[1:]\n",
    "            prof_list.append(array)\n",
    "            long+=1\n",
    "    return(prof_list,long)\n",
    "\n",
    "def add_window_spaces(prof_list,window):\n",
    "    for i in range(window//2):\n",
    "        prof_list=[[0.0 for i in range(20)]]+prof_list\n",
    "        prof_list.append([0.0 for i in range(20)])\n",
    "    prof_list=np.array([np.array(x) for x in prof_list],dtype=float)\n",
    "    return(prof_list)\n",
    "\n",
    "def prediction(window_list,info_list_fun):  \n",
    "    res=[]\n",
    "    for ar in info_list_fun:\n",
    "        tot=0\n",
    "        for li in range(len(ar[1])):\n",
    "            for el in range(len(ar[1][li])):\n",
    "                tot=tot+(ar[1][li][el]*window_list[li][el])\n",
    "        res.append([ar[0],tot])\n",
    "    #print(res)\n",
    "    maxi=res[0][1]\n",
    "    if res[0][1]==0.0 and res[1][1]==0.0 and res[2][1]==0.0:\n",
    "        return(\"C\")\n",
    "    else:\n",
    "        for el in range(len(res)):\n",
    "            if res[el][1]>=maxi:\n",
    "                maxi=res[el][1]\n",
    "                maxi_ss=res[el][0]\n",
    "        return(maxi_ss)\n",
    "\n",
    "prediction_df=pd.DataFrame(columns=[\"Predicted Secondary Structure\"])\n",
    "for prof_name in os.listdir(prof_path):\n",
    "    #print(\"\\n\", prof_name)\n",
    "    prof_name=prof_name.split(\".\")\n",
    "    prof_name=\".\".join(prof_name[:-1])\n",
    "    profile=open(prof_path+prof_name+\".pssm\")\n",
    "    prof_list,long=reading_profile(profile) #function recall\n",
    "    prof_ar=add_window_spaces(prof_list,window) #function recall\n",
    "    predicted=\"\"\n",
    "    for el in range(long):\n",
    "        window_list=prof_ar[el:el+window]\n",
    "        predict_ss=prediction(window_list,info_list_fun[:-1]) #function recall\n",
    "        predicted=predicted+predict_ss\n",
    "    #print(prof_name)\n",
    "    #print(predicted)\n",
    "    df_row=pd.Series({\"Predicted Secondary Structure\":predicted},name=prof_name)\n",
    "    prediction_df=prediction_df.append(df_row)\n",
    "prediction_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Support Vector Machines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will use the **scikit-lear** library. https://scikit-learn.org/stable/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SVMs with scikit-learn: https://scikit-learn.org/stable/modules/svm.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Different implementatios for different problems:\n",
    "- SVC: Support Vector Classification\n",
    "- NuSVC: A different mathematical formulation of SVC\n",
    "- SVR: Support vector Regre\n",
    "\n",
    "We will focus on multi-class SVC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Training data consist of:\n",
    "- A matrix **X** contining all example vectors (in rows)\n",
    "- A vector **y** containing target classes (on for each vector in x)\n",
    "- **The number of rows in X = the number of elements in y**\n",
    "\n",
    "I have two options:\n",
    "\n",
    "- Define **X** as a list of floats and **y** as a list of integers\n",
    "- Define both **X** and **y** as **numpy** arrays (**X** is bi-dimensional whereas **y** is unidimesional.\n",
    "\n",
    "In both cases I have to use a **dense representation**.\n",
    "- Also **zero frequences are reported**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SVM for SS prediction\n",
    "\n",
    "Our **input vectors** are **sequence profile windows**\n",
    "\n",
    "We can represent one profile **window of size w** as a **vector of dimension w$\\times$20**\n",
    "\n",
    "**Each window is a pont in the w$\\times$20 dimensional space**\n",
    "\n",
    "Window vectors belong to one of the three different classes assigned according to the conformation of the central residue: **1(Helix), 2 (Strand) or 3 (Coil)**\n",
    "\n",
    "**Multi-class SVM** implemented by LIBSVM (and scikit-learn too) using the **one-against-rest** aproach:\n",
    "- **Three binary SVMs** are actually trained: **Helix vs Other** and **Coil vs Other**\n",
    "- In the prediction phase, the predicted class is chosen according to the value of the discrimination function for the three SVMs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will now extract and create the input for the SVC process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_path=\"JPred_pssm/\"\n",
    "dssp_path=\"JPred_dssp/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "198803\n",
      "198803\n",
      "CPU times: user 7.97 s, sys: 639 ms, total: 8.61 s\n",
      "Wall time: 9.21 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "window=17\n",
    "def reading_files(dssp,profile):\n",
    "    profile_list=[]\n",
    "    for line in dssp:\n",
    "        if line.startswith(\">\"):\n",
    "            sec_str1=\"\"\n",
    "            sec_str=dssp.readline().rstrip()\n",
    "            for el in sec_str: \n",
    "                if el==\"-\":\n",
    "                    sec_str1=sec_str1+\"C\"\n",
    "                else:\n",
    "                    sec_str1=sec_str1+el \n",
    "    for line in profile:\n",
    "        if \"A\" not in line:\n",
    "            line=line.rstrip()\n",
    "            line=line.split()[1:]\n",
    "            profile_list.append(line)\n",
    "    profile_list=[[float(j) for j in i] for i in profile_list] #COOL\n",
    "    return(sec_str1,profile_list)\n",
    "\n",
    "def class_assign(sec_str):\n",
    "    ss_class=[]\n",
    "    for ss in sec_str:\n",
    "        if ss==\"H\":\n",
    "            ss_class.append(1)\n",
    "        if ss==\"E\":\n",
    "            ss_class.append(2)\n",
    "        if ss==\"C\":\n",
    "            ss_class.append(3)\n",
    "    return(ss_class)\n",
    "\n",
    "def add_window_spaces(profile_list,window):\n",
    "    for i in range(window//2):\n",
    "        profile_list=[[0.0 for i in range(20)]]+profile_list\n",
    "        profile_list.append([0.0 for i in range(20)])\n",
    "    return(profile_list)\n",
    "\n",
    "y=[]\n",
    "X=[]\n",
    "for filename in os.listdir(profile_path):\n",
    "    filename=filename.split(\".\")\n",
    "    filename=\".\".join(filename[:-1])\n",
    "    profile=open(profile_path+filename+\".pssm\")\n",
    "    dssp=open(dssp_path+filename+\".dssp\")\n",
    "    sec_str,profile_list=reading_files(dssp,profile) #function recall\n",
    "    ss_class=class_assign(sec_str) #function recall\n",
    "    profile_list=add_window_spaces(profile_list,window) #function recall\n",
    "    for el in range(len(ss_class)):\n",
    "        window_list=profile_list[el:el+window]\n",
    "        window_list=[el for li in window_list for el in li]\n",
    "        y.append(ss_class[el])\n",
    "        X.append(window_list)\n",
    "print(len(y))\n",
    "print(len(X))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now with the python scritp present in the folder is possible to perform the 5-fold cross validation on the entire JPred4 dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation procedure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Evaluating binary classifiers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Main binary scoring indexes:\n",
    "\n",
    "$$ACC=\\frac{TP+TN}{TP+TN+FP+FN}$$\n",
    "\n",
    "$$Sen=\\frac{TP}{TP+FN}$$\n",
    "\n",
    "$$PPV=\\frac{TP}{TP+FP}$$\n",
    "\n",
    "$$MCC=\\frac{TP\\times TN-FP\\times FN}{\\sqrt{(TP+FP)\\times(TP+FN)\\times(TN+FP)\\times(TN+FN)}}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Multi-class problems\n",
    "\n",
    "**K-calss confusion matrix**\n",
    "\n",
    "$p_{ij}$=object in class i predicted in class j"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$Q_k=\\frac{p_{11}+p_{22}+...+p_{KK}}{N}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Binary scores computed from the binary matrix e.g. foe helix\n",
    "\n",
    "$$Sen_h=\\frac{c_H}{c_H+c_H}$$\n",
    "\n",
    "$$PPV_H=\\frac{c_H}{c_H+o_H}$$\n",
    "\n",
    "$$MCC=\\frac{c_H\\times n_H-o_H\\times u_H}{\\sqrt{(c_H+o_H)\\times(c_H+u_H)\\times(n_H+o_H)\\times(n_H+u_H)}}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- $c_H$ = $p_{HH}$ = helix resdue correctly predicted in class H (**correct positive**)\n",
    "- $o_H$ = $p_{EH}+p_{CH}$ = non-helix residues predicted as H (**over-predictions**)\n",
    "- $u_H$ = $p_{HE}+p_{HC}$ = helix residues as non-H (**under-prediction**)\n",
    "- $n_H$ = $p_{EE}+p_{CE}+p_{EC}+p_{CC}$ = non-helix residues predicted as non.-H (**correct negative**)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Three-class accuracy ($Q_3$) is computed from 3-class matrix\n",
    "\n",
    "$$Q_3=\\frac{p_{HH}+p_{EE}+p_{CC}}{N}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will be the evaluation of the GOR model trained on the entire JPred class and then tested on the Blind_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "blind_dssp=\"Blindset_DSSP/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy:  0.6377463171991582\n",
      "- Sensitivity \t for H: 0.7825829383886256\tfor E: 0.6917435037720033\tfor C: 0.44542619542619544\n",
      "- Precision \t for H: 0.6642196299275945\tfor E: 0.5228064618308521\tfor C: 0.7372043010752688\n",
      "- MCC \t\t for H: 0.5057361334430451\tfor E: 0.4616530439280952\tfor C: 0.4093086432120267\n"
     ]
    }
   ],
   "source": [
    "conf_dict={}\n",
    "\n",
    "def reading_files(obs_file):\n",
    "    for line in obs_file:\n",
    "        if line.startswith(\">\"):\n",
    "            obs=obs_file.readline().rstrip()\n",
    "            return(obs)\n",
    "\n",
    "def conf_mat(pred,obs,conf_dict):\n",
    "    for ss in range(len(obs)):\n",
    "        key=conf_dict.keys()\n",
    "        if obs[ss]+pred[ss] not in key:\n",
    "            conf_dict[obs[ss]+pred[ss]]=1\n",
    "        else: \n",
    "            conf_dict[obs[ss]+pred[ss]]+=1\n",
    "    return(conf_dict)\n",
    "\n",
    "def accuracy(conf_dict):\n",
    "    tot=0\n",
    "    for p in conf_dict:\n",
    "        tot+=conf_dict[p]\n",
    "    acc=(conf_dict[\"HH\"]+conf_dict[\"EE\"]+conf_dict[\"CC\"])/tot\n",
    "    return(acc)\n",
    "\n",
    "def sensitivity(conf_dict):\n",
    "    sen_H=conf_dict[\"HH\"]/(conf_dict[\"HH\"]+(conf_dict[\"HE\"]+conf_dict[\"HC\"])) #conf_dict[\"HE\"] --> observed as helixes (H) and predicted ass beta-strand or Coil (P[\"HC\"]) \n",
    "    sen_E=conf_dict[\"EE\"]/(conf_dict[\"EE\"]+(conf_dict[\"EH\"]+conf_dict[\"EC\"]))\n",
    "    sen_C=conf_dict[\"CC\"]/(conf_dict[\"CC\"]+(conf_dict[\"CE\"]+conf_dict[\"CH\"]))\n",
    "    return(sen_H,sen_E,sen_C)\n",
    "    \n",
    "def precision(cond_dict):\n",
    "    ppv_H=conf_dict[\"HH\"]/(conf_dict[\"HH\"]+(conf_dict[\"EH\"]+conf_dict[\"CH\"])) #conf_dict[\"EH\"]+conf_dict[\"CH\"] --> non-helix residues predicted as H\n",
    "    ppv_E=conf_dict[\"EE\"]/(conf_dict[\"EE\"]+(conf_dict[\"HE\"]+conf_dict[\"CE\"]))     \n",
    "    ppv_C=conf_dict[\"CC\"]/(conf_dict[\"CC\"]+(conf_dict[\"EC\"]+conf_dict[\"HC\"]))\n",
    "    return(ppv_H,ppv_E,ppv_C)\n",
    "\n",
    "def MCC(cond_dict):\n",
    "    mcc_H=(cond_dict[\"HH\"]*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"])-((cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"HE\"]+cond_dict[\"HC\"])))/math.sqrt((cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"HH\"]+cond_dict[\"HE\"]+cond_dict[\"HC\"])*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"]+cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"]+cond_dict[\"HE\"]+cond_dict[\"HC\"]))\n",
    "    mcc_E=(cond_dict[\"EE\"]*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"])-((cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"EH\"]+cond_dict[\"EC\"])))/math.sqrt((cond_dict[\"EE\"]+cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"EE\"]+cond_dict[\"EH\"]+cond_dict[\"EC\"])*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"]+cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"]+cond_dict[\"EH\"]+cond_dict[\"EC\"]))\n",
    "    mcc_C=(cond_dict[\"CC\"]*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"])-((cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"CH\"]+cond_dict[\"CE\"])))/math.sqrt((cond_dict[\"CC\"]+cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"CC\"]+cond_dict[\"CH\"]+cond_dict[\"CE\"])*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"]+cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"]+cond_dict[\"CH\"]+cond_dict[\"CE\"]))\n",
    "    return(mcc_H,mcc_E,mcc_C)\n",
    "\n",
    "for index in prediction_df.index:\n",
    "    pred=prediction_df.loc[index,\"Predicted Secondary Structure\"]\n",
    "    obs_file=open(blind_dssp+index+\".dssp\")\n",
    "    obs=reading_files(obs_file) #function recall\n",
    "    conf_dict=conf_mat(pred,obs,conf_dict) #function recall\n",
    "#print(conf_dict)    \n",
    "acc=accuracy(conf_dict)\n",
    "sen_H,sen_E,sen_C=sensitivity(conf_dict)\n",
    "ppv_H,ppv_E,ppv_C=precision(conf_dict)\n",
    "mcc_H,mcc_E,mcc_C=MCC(conf_dict)\n",
    "print(\"Accuracy: \",acc)\n",
    "print(\"- Sensitivity\",\"\\t\", \"for H:\",sen_H , end=\"\\t\")\n",
    "print(\"for E:\",sen_E, end=\"\\t\")\n",
    "print(\"for C:\",sen_C)\n",
    "print(\"- Precision\",\"\\t\", \"for H:\", ppv_H , end=\"\\t\")\n",
    "print(\"for E:\",ppv_E, end=\"\\t\")\n",
    "print(\"for C:\",ppv_C)\n",
    "print(\"- MCC\" ,\"\\t\\t\" ,\"for H:\", mcc_H , end=\"\\t\")\n",
    "print(\"for E:\",mcc_E, end=\"\\t\")\n",
    "print(\"for C:\",mcc_C)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5-fold cross validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Gor-Matrix**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Gor_method1(pssm_path,dssp_path,train_set,pssm_files):\n",
    "    def extracter(dssp,pssm):\n",
    "        for line in dssp:\n",
    "            if line.startswith(\">\"):\n",
    "                sec_str1=\"\"\n",
    "                sec_str=dssp.readline().rstrip()\n",
    "                for el in sec_str: \n",
    "                    if el==\"-\":\n",
    "                        sec_str1=sec_str1+\"C\"\n",
    "                    else:\n",
    "                        sec_str1=sec_str1+el \n",
    "        for line in pssm:\n",
    "            if \"A\" not in line:\n",
    "                line=line.rstrip()\n",
    "                line=line.split()[1:]\n",
    "                pssm_list.append(line)\n",
    "        return(sec_str1,pssm_list)\n",
    "\n",
    "    def window_pssm(pssm_list,sec_str,window,el):\n",
    "            window_array=pssm_list[el-(window//2):el+window//2+1]\n",
    "            return(window_array,sec_str[el-window//2])\n",
    "\n",
    "    def add_window_spaces(pssm_list,window):\n",
    "        for i in range(window//2):\n",
    "            pssm_list=[[0.0 for i in range(20)]]+pssm_list\n",
    "            pssm_list.append([0.0 for i in range(20)])\n",
    "        pssm_array=np.array([np.array(list) for list in pssm_list])\n",
    "        pssm_list_float=pssm_array.astype(np.float)\n",
    "        return(pssm_list_float)\n",
    "\n",
    "    def fill_gor_mat(window_array,el_sec_str,gor_list_mat):\n",
    "        for list in range(len(gor_list_mat)):\n",
    "            if gor_list_mat[list][0]==el_sec_str:\n",
    "                gor_list_mat[list][1]+=window_array\n",
    "        gor_list_mat[3][1]+=window_array\n",
    "        return(gor_list_mat)\n",
    "\n",
    "    def gor_df_fun(tot_gor_mat,secstr,aa,window):\n",
    "        index=[]\n",
    "        gor_df1=pd.DataFrame(tot_gor_mat)\n",
    "        gor_df1.columns=aa\n",
    "        for el in secstr:\n",
    "            for i in range(window):\n",
    "                index.append(\"R\"+str(i-(window//2))+\"_\"+el)\n",
    "        gor_df1.index=index\n",
    "        return(gor_df1)\n",
    "\n",
    "    def normalization(gor_list_mat):\n",
    "        summation=[]\n",
    "        summation=np.sum(gor_list_mat[3][1],axis=1)\n",
    "        for array in gor_list_mat:\n",
    "            for sub in range(len(array[1])):\n",
    "                array[1][sub]=array[1][sub]/summation[sub]\n",
    "        return(gor_list_mat)\n",
    "\n",
    "\n",
    "    aa=['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "    secstr=[\"H\",\"E\",\"C\",\"tot\"]\n",
    "    window=17\n",
    "    gor_list_mat=[]\n",
    "    for el in secstr:\n",
    "        gor_list_mat.append([el,np.zeros(((window),len(aa)),float)])\n",
    "    prob_sec_str={}\n",
    "    \n",
    "    for filename in train_set:\n",
    "        #print(pssm_file)\n",
    "        if filename not in pssm_files:\n",
    "            continue\n",
    "        pssm_list=[]\n",
    "        pssm=open(pssm_path+filename+\".pssm\")\n",
    "        dssp=open(dssp_path+filename+\".dssp\")\n",
    "        sec_str,pssm_list=extracter(dssp,pssm) # function recall\n",
    "        pssm_array=add_window_spaces(pssm_list,window) #function recall\n",
    "        for el in range(window//2,len(sec_str)+window//2):\n",
    "            window_array,el_sec_str=window_pssm(pssm_array,sec_str,window,el) #function recall\n",
    "            gor_list_mat=fill_gor_mat(window_array,el_sec_str,gor_list_mat) #function recall\n",
    "            keys=prob_sec_str.keys()\n",
    "            if el_sec_str in keys:\n",
    "                prob_sec_str[el_sec_str]+=1\n",
    "            else: \n",
    "                prob_sec_str[el_sec_str]=1\n",
    "    gor_list_mat=normalization(gor_list_mat) #function recall\n",
    "    tot_gor_mat=np.concatenate((gor_list_mat[0][1],gor_list_mat[1][1],gor_list_mat[2][1],gor_list_mat[3][1]),axis=0)\n",
    "    gor_df=gor_df_fun(tot_gor_mat,secstr,aa,window) #fucntion recall\n",
    "    #print(\"Number of the secondary structures: \",prob_sec_str)\n",
    "    return(gor_df,prob_sec_str,gor_list_mat)\n",
    "\n",
    "#gor_df,prob_sec_str,gor_list_mat=Gor_method1(pssm_path,dssp_path,train_set)\n",
    "#print(gor_list_mat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Information function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def info_matrix(prob_sec_str,gor_list_mat):\n",
    "    tot_sec_str=0\n",
    "    for el in prob_sec_str:\n",
    "        tot_sec_str+=prob_sec_str[el]\n",
    "    for el in prob_sec_str:\n",
    "        prob_sec_str[el]=prob_sec_str[el]/tot_sec_str\n",
    "    #print(prob_sec_str)\n",
    "    \n",
    "    info_list_fun=gor_list_mat\n",
    "    \n",
    "    #print(info_list_fun)\n",
    "    for mat in info_list_fun:\n",
    "        if mat[0]==\"tot\":\n",
    "            tot_info_fun=mat[1]\n",
    "\n",
    "    for ar in (info_list_fun[:-1]):\n",
    "        for vec in range(len(ar[1])):\n",
    "            for el in range(len(ar[1][vec])):\n",
    "                P_ss=prob_sec_str[ar[0]]\n",
    "                P_join=ar[1][vec][el]\n",
    "                P_res=tot_info_fun[vec][el]\n",
    "                ar[1][vec][el]=math.log2(P_join/(P_ss*P_res))\n",
    "    #print(info_list_fun)\n",
    "    return(info_list_fun)\n",
    "    \n",
    "#info_list_fun=info_matrix(prob_sec_str,gor_list_mat)\n",
    "#print(info_list_fun)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Prediction**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prediction1(jpred_pssm_path,test_set,pssm_files,info_list_fun):\n",
    "    #import math\n",
    "    def reading_profile(profile):\n",
    "        prof_list=[]\n",
    "        long=0\n",
    "        for line in profile:\n",
    "            if \"A\" not in line:\n",
    "                #prof_line=[]\n",
    "                line=line.rstrip()\n",
    "                array=line.split()[1:]\n",
    "                prof_list.append(array)\n",
    "                long+=1\n",
    "        return(prof_list,long)\n",
    "\n",
    "    def add_window_spaces(prof_list,window):\n",
    "        for i in range(window//2):\n",
    "            prof_list=[[0.0 for i in range(20)]]+prof_list\n",
    "            prof_list.append([0.0 for i in range(20)])\n",
    "        prof_list=np.array([np.array(x) for x in prof_list],dtype=float)\n",
    "        return(prof_list)\n",
    "\n",
    "    def prediction(window_list,info_list_fun):  \n",
    "        res=[]\n",
    "        for ar in info_list_fun:\n",
    "            tot=0\n",
    "            for li in range(len(ar[1])):\n",
    "                for el in range(len(ar[1][li])):\n",
    "                    tot=tot+(ar[1][li][el]*window_list[li][el])\n",
    "            res.append([ar[0],tot])\n",
    "        #print(res)\n",
    "        maxi=res[0][1]\n",
    "        if res[0][1]==0.0 and res[1][1]==0.0 and res[2][1]==0.0:\n",
    "            return(\"C\")\n",
    "        else:\n",
    "            for el in range(len(res)):\n",
    "                if res[el][1]>=maxi:\n",
    "                    maxi=res[el][1]\n",
    "                    maxi_ss=res[el][0]\n",
    "            return(maxi_ss)\n",
    "\n",
    "    #print(pssm_files)\n",
    "    prediction_df=pd.DataFrame(columns=[\"Predicted Secondary Structure\"])\n",
    "    window=17\n",
    "    for filename in test_set:\n",
    "        #print(filename)\n",
    "        if filename not in pssm_files:\n",
    "            continue\n",
    "        profile=open(jpred_pssm_path+filename+\".pssm\")\n",
    "        prof_list,long=reading_profile(profile) #function recall\n",
    "        prof_ar=add_window_spaces(prof_list,window) #function recall\n",
    "        predicted=\"\"\n",
    "        for el in range(long):\n",
    "            window_list=prof_ar[el:el+window]\n",
    "            predict_ss=prediction(window_list,info_list_fun[:-1]) #function recall\n",
    "            predicted=predicted+predict_ss\n",
    "        #print(prof_name)\n",
    "        #print(predicted)\n",
    "        df_row=pd.Series({\"Predicted Secondary Structure\":predicted},name=filename)\n",
    "        prediction_df=prediction_df.append(df_row)\n",
    "    #print(prediction_df)\n",
    "    return(prediction_df)\n",
    "\n",
    "#prediction_df=prediction1(jpred_pssm_path,test_set,pssm_files,info_list_fun)\n",
    "#prediction_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Evaluation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluation1(prediction_df,jpred_dssp_path):\n",
    "    \n",
    "    def reading_files(obs_file):\n",
    "        for line in obs_file:\n",
    "            if line.startswith(\">\"):\n",
    "                obs=obs_file.readline().rstrip()\n",
    "                return(obs)\n",
    "\n",
    "    def conf_mat(pred,obs,conf_dict):\n",
    "        for ss in range(len(obs)):\n",
    "            key=conf_dict.keys()\n",
    "            if obs[ss]+pred[ss] not in key:\n",
    "                conf_dict[obs[ss]+pred[ss]]=1\n",
    "            else: \n",
    "                conf_dict[obs[ss]+pred[ss]]+=1\n",
    "        return(conf_dict)\n",
    "\n",
    "    def accuracy(conf_dict):\n",
    "        tot=0\n",
    "        for p in conf_dict:\n",
    "            tot+=conf_dict[p]\n",
    "        #print(tot)\n",
    "        acc=(conf_dict[\"HH\"]+conf_dict[\"EE\"]+conf_dict[\"CC\"])/tot\n",
    "        return(acc)\n",
    "\n",
    "    def sensitivity(conf_dict):\n",
    "        sen_H=conf_dict[\"HH\"]/(conf_dict[\"HH\"]+(conf_dict[\"HE\"]+conf_dict[\"HC\"])) #conf_dict[\"HE\"] --> observed as helixes (H) and predicted ass beta-strand or Coil (P[\"HC\"]) \n",
    "        sen_E=conf_dict[\"EE\"]/(conf_dict[\"EE\"]+(conf_dict[\"EH\"]+conf_dict[\"EC\"]))\n",
    "        sen_C=conf_dict[\"CC\"]/(conf_dict[\"CC\"]+(conf_dict[\"CE\"]+conf_dict[\"CH\"]))\n",
    "        return(sen_H,sen_E,sen_C)\n",
    "\n",
    "    def precision(cond_dict):\n",
    "        ppv_H=conf_dict[\"HH\"]/(conf_dict[\"HH\"]+(conf_dict[\"EH\"]+conf_dict[\"CH\"])) #conf_dict[\"EH\"]+conf_dict[\"CH\"] --> non-helix residues predicted as H\n",
    "        ppv_E=conf_dict[\"EE\"]/(conf_dict[\"EE\"]+(conf_dict[\"HE\"]+conf_dict[\"CE\"]))     \n",
    "        ppv_C=conf_dict[\"CC\"]/(conf_dict[\"CC\"]+(conf_dict[\"EC\"]+conf_dict[\"HC\"]))\n",
    "        return(ppv_H,ppv_E,ppv_C)\n",
    "\n",
    "    def MCC(cond_dict):\n",
    "        mcc_H=(cond_dict[\"HH\"]*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"])-((cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"HE\"]+cond_dict[\"HC\"])))/math.sqrt((cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"HH\"]+cond_dict[\"HE\"]+cond_dict[\"HC\"])*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"]+cond_dict[\"EH\"]+cond_dict[\"CH\"])*(cond_dict[\"EE\"]+cond_dict[\"CE\"]+cond_dict[\"EC\"]+cond_dict[\"CC\"]+cond_dict[\"HE\"]+cond_dict[\"HC\"]))\n",
    "        mcc_E=(cond_dict[\"EE\"]*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"])-((cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"EH\"]+cond_dict[\"EC\"])))/math.sqrt((cond_dict[\"EE\"]+cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"EE\"]+cond_dict[\"EH\"]+cond_dict[\"EC\"])*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"]+cond_dict[\"HE\"]+cond_dict[\"CE\"])*(cond_dict[\"HH\"]+cond_dict[\"CH\"]+cond_dict[\"HC\"]+cond_dict[\"CC\"]+cond_dict[\"EH\"]+cond_dict[\"EC\"]))\n",
    "        mcc_C=(cond_dict[\"CC\"]*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"])-((cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"CH\"]+cond_dict[\"CE\"])))/math.sqrt((cond_dict[\"CC\"]+cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"CC\"]+cond_dict[\"CH\"]+cond_dict[\"CE\"])*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"]+cond_dict[\"HC\"]+cond_dict[\"EC\"])*(cond_dict[\"HH\"]+cond_dict[\"EH\"]+cond_dict[\"HE\"]+cond_dict[\"EE\"]+cond_dict[\"CH\"]+cond_dict[\"CE\"]))\n",
    "        return(mcc_H,mcc_E,mcc_C)\n",
    "\n",
    "    conf_dict={}\n",
    "    for index in prediction_df.index:\n",
    "        pred=prediction_df.loc[index,\"Predicted Secondary Structure\"]\n",
    "        obs_file=open(jpred_dssp_path+index+\".dssp\")\n",
    "        obs=reading_files(obs_file) #function recall\n",
    "        conf_dict=conf_mat(pred,obs,conf_dict) #function recall\n",
    "    #print(conf_dict)  \n",
    "    for key in conf_dict.copy():\n",
    "        if \"-\" in key:\n",
    "            new_key=key.replace(\"-\",\"C\")\n",
    "            conf_dict[new_key]=conf_dict.pop(key)\n",
    "    #print(conf_dict)\n",
    "    acc=accuracy(conf_dict)\n",
    "    sen_H,sen_E,sen_C=sensitivity(conf_dict)\n",
    "    ppv_H,ppv_E,ppv_C=precision(conf_dict)\n",
    "    mcc_H,mcc_E,mcc_C=MCC(conf_dict)\n",
    "    print(\"Accuracy: \",acc)\n",
    "    print(\"- Sensitivity\",\"\\t\", \"for H:\",sen_H , end=\"\\t\")\n",
    "    print(\"for E:\",sen_E, end=\"\\t\")\n",
    "    print(\"for C:\",sen_C)\n",
    "    print(\"- Precision\",\"\\t\", \"for H:\", ppv_H , end=\"\\t\")\n",
    "    print(\"for E:\",ppv_E, end=\"\\t\")\n",
    "    print(\"for C:\",ppv_C)\n",
    "    print(\"- MCC\" ,\"\\t\\t\" ,\"for H:\", mcc_H , end=\"\\t\")\n",
    "    print(\"for E:\",mcc_E, end=\"\\t\")\n",
    "    print(\"for C:\",mcc_C)\n",
    "    return(sen_H,sen_E,sen_C,ppv_H,ppv_E,ppv_C,mcc_H,mcc_E,mcc_C,acc)\n",
    "    \n",
    "#evaluation1(prediction_df,jpred_dssp_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now That I have defined all these 4 function I can calculate each of the evaluation parameters for all the 5 cross validation (set0, set1, set2, set3, set4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tested on:  set0\n",
      "Accuracy:  0.6251482506424194\n",
      "- Sensitivity \t for H: 0.7847054770926628\tfor E: 0.7161552346570397\tfor C: 0.4424618264786755\n",
      "- Precision \t for H: 0.6424864620938628\tfor E: 0.4784083201446982\tfor C: 0.7982058047493403\n",
      "- MCC \t\t for H: 0.5225095376245243\tfor E: 0.43803593271084296\tfor C: 0.4207203932944134\n",
      "tested on:  set1\n",
      "Accuracy:  0.6272471091244777\n",
      "- Sensitivity \t for H: 0.8077864133563616\tfor E: 0.6978571428571428\tfor C: 0.44401190748797803\n",
      "- Precision \t for H: 0.6154731878495449\tfor E: 0.514442605686776\tfor C: 0.8052325581395349\n",
      "- MCC \t\t for H: 0.5241478638564095\tfor E: 0.44816697667458694\tfor C: 0.425922372432941\n",
      "tested on:  set2\n",
      "Accuracy:  0.626651747655584\n",
      "- Sensitivity \t for H: 0.809719370294319\tfor E: 0.7308157831033648\tfor C: 0.424\n",
      "- Precision \t for H: 0.6367061356297093\tfor E: 0.48673921805723497\tfor C: 0.8130574384587942\n",
      "- MCC \t\t for H: 0.5381030668227689\tfor E: 0.4521345798005523\tfor C: 0.4162317131669378\n",
      "tested on:  set3\n",
      "Accuracy:  0.6254109850898433\n",
      "- Sensitivity \t for H: 0.7989993915218714\tfor E: 0.6951830443159923\tfor C: 0.43868179362506754\n",
      "- Precision \t for H: 0.6501265265705798\tfor E: 0.45740365111561865\tfor C: 0.7921951219512195\n",
      "- MCC \t\t for H: 0.5236625866226363\tfor E: 0.4267138801131944\tfor C: 0.41233582966861393\n",
      "tested on:  set4\n",
      "Accuracy:  0.6229577185859986\n",
      "- Sensitivity \t for H: 0.809842769512797\tfor E: 0.6923658352229781\tfor C: 0.42868098159509205\n",
      "- Precision \t for H: 0.6208313064158694\tfor E: 0.5028625205866206\tfor C: 0.7947287839020123\n",
      "- MCC \t\t for H: 0.519462079715843\tfor E: 0.4420549443815164\tfor C: 0.41112031260556536\n",
      "\n",
      "Total mean and standard deviation for all the three secondary structures:\n",
      "\n",
      "Accuracy: 0.6254831622196646 ± 0.0016562262416038467\n",
      "\n",
      "Sensitivity:\t for H 0.8022106843556024 ± 0.004808651189498475 \t for E 0.7064754080313036 ± 0.00737232975346802 \t for C 0.43556730183736264 ± 0.003935541716177811\n",
      "Precision:\t for H 0.6331247237119132 ± 0.0065277667866472915 \t for E 0.48797126311818967 ± 0.009873773957461121 \t for C 0.8006839414401803 ± 0.0037911638148881172\n",
      "MCC:\t \t for H 0.5255770269284364 ± 0.0032358255070981824 \t for E 0.4414212627361386 ± 0.004407090505580596 \t for C 0.4172661242336943 ± 0.0027387170258420038\n"
     ]
    }
   ],
   "source": [
    "#import statistics\n",
    "jpred_pssm_path=\"JPred_pssm/\"\n",
    "cross_val_path=\"CV/\"\n",
    "jpred_dssp_path=\"JPred_dssp/\"\n",
    "\n",
    "sen_tot_H,sen_tot_E,sen_tot_C=[],[],[]\n",
    "ppv_tot_H,ppv_tot_E,ppv_tot_C=[],[],[]\n",
    "mcc_tot_H,mcc_tot_E,mcc_tot_C=[],[],[]\n",
    "accuracy_tot=[]\n",
    "\n",
    "pssm_files=[]\n",
    "for pssm_file in os.listdir(jpred_pssm_path):\n",
    "    pssm_files.append(pssm_file[:-5])\n",
    "        \n",
    "cv_set=[]\n",
    "for file_name in os.listdir(cross_val_path):\n",
    "    cv_set.append(file_name)\n",
    "for test in cv_set:\n",
    "    test_set=[]\n",
    "    test_ids=open(cross_val_path+test)\n",
    "    for line in test_ids:\n",
    "        line=line.rstrip()\n",
    "        test_set.append(line)\n",
    "    train_set=[]\n",
    "    for train in cv_set:\n",
    "        if train!=test:\n",
    "            train_ids=open(cross_val_path+train)\n",
    "            for line in train_ids:\n",
    "                line=line.rstrip()\n",
    "                train_set.append(line) \n",
    "    print(\"tested on: \",test)\n",
    "    gor_df,prob_sec_str,gor_list_mat=Gor_method1(jpred_pssm_path,jpred_dssp_path,train_set,pssm_files)\n",
    "    info_list_fun=info_matrix(prob_sec_str,gor_list_mat)\n",
    "    prediction_df=prediction1(jpred_pssm_path,test_set,pssm_files,info_list_fun)\n",
    "    #print(len(prediction_df))\n",
    "    #evaluation1(prediction_df,jpred_dssp_path)\n",
    "    sen_H,sen_E,sen_C,ppv_H,ppv_E,ppv_C,mcc_H,mcc_E,mcc_C,acc=evaluation1(prediction_df,jpred_dssp_path)\n",
    "    sen_tot_H.append(sen_H)\n",
    "    sen_tot_E.append(sen_E)\n",
    "    sen_tot_C.append(sen_C)\n",
    "    ppv_tot_H.append(ppv_H)\n",
    "    ppv_tot_E.append(ppv_E)\n",
    "    ppv_tot_C.append(ppv_C)\n",
    "    mcc_tot_H.append(mcc_H)\n",
    "    mcc_tot_E.append(mcc_E)\n",
    "    mcc_tot_C.append(mcc_C)\n",
    "    accuracy_tot.append(acc)\n",
    "print()\n",
    "print(\"Total mean and standard deviation for all the three secondary structures:\")\n",
    "print()\n",
    "print(\"Accuracy:\",statistics.mean(accuracy_tot),\"±\",statistics.stdev(accuracy_tot))\n",
    "print()\n",
    "print(\"Sensitivity:\\t for H\",statistics.mean(sen_tot_H),\"±\",statistics.stdev(sen_tot_H)/math.sqrt(5),\"\\t for E\",statistics.mean(sen_tot_E),\"±\",statistics.stdev(sen_tot_E)/math.sqrt(5),\"\\t for C\",statistics.mean(sen_tot_C),\"±\",statistics.stdev(sen_tot_C)/math.sqrt(5))\n",
    "print(\"Precision:\\t for H\",statistics.mean(ppv_tot_H),\"±\",statistics.stdev(ppv_tot_H)/math.sqrt(5),\"\\t for E\",statistics.mean(ppv_tot_E),\"±\",statistics.stdev(ppv_tot_E)/math.sqrt(5),\"\\t for C\",statistics.mean(ppv_tot_C),\"±\",statistics.stdev(ppv_tot_C)/math.sqrt(5))\n",
    "print(\"MCC:\\t \\t for H\",statistics.mean(mcc_tot_H),\"±\",statistics.stdev(mcc_tot_H)/math.sqrt(5),\"\\t for E\",statistics.mean(mcc_tot_E),\"±\",statistics.stdev(mcc_tot_E)/math.sqrt(5),\"\\t for C\",statistics.mean(mcc_tot_C),\"±\",statistics.stdev(mcc_tot_C)/math.sqrt(5))    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Other used scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_path=\"JPred_pssm/\"\n",
    "dssp_path=\"JPred_dssp/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test 40396 40396\n",
      "train 158407 158407\n"
     ]
    }
   ],
   "source": [
    "jpred_pssm_path=\"JPred_pssm/\"\n",
    "cross_val_path=\"CV/\"\n",
    "jpred_dssp_path=\"JPred_dssp/\"\n",
    "\n",
    "def input_SVM(test_train_set,pssm_files,jpred_pssm_path,jpred_dssp_path):\n",
    "    window=17\n",
    "    def reading_files(dssp,profile):\n",
    "        profile_list=[]\n",
    "        for line in dssp:\n",
    "            if line.startswith(\">\"):\n",
    "                sec_str1=\"\"\n",
    "                sec_str=dssp.readline().rstrip()\n",
    "                for el in sec_str: \n",
    "                    if el==\"-\":\n",
    "                        sec_str1=sec_str1+\"C\"\n",
    "                    else:\n",
    "                        sec_str1=sec_str1+el \n",
    "                for line in profile:\n",
    "                    if \"A\" not in line:\n",
    "                        line=line.rstrip()\n",
    "                        line=line.split()[1:]\n",
    "                        profile_list.append(line)\n",
    "        profile_list=[[float(j) for j in i] for i in profile_list] #COOL\n",
    "        return(sec_str1,profile_list)\n",
    "\n",
    "    def class_assign(sec_str):\n",
    "        ss_class=[]\n",
    "        for ss in sec_str:\n",
    "            if ss==\"H\":\n",
    "                ss_class.append(1)\n",
    "            if ss==\"E\":\n",
    "                ss_class.append(2)\n",
    "            if ss==\"C\":\n",
    "                ss_class.append(3)\n",
    "        return(ss_class)\n",
    "\n",
    "    def add_window_spaces(profile_list,window):\n",
    "        for i in range(window//2):\n",
    "            profile_list=[[0.0 for i in range(20)]]+profile_list\n",
    "            profile_list.append([0.0 for i in range(20)])\n",
    "        return(profile_list)\n",
    "\n",
    "    y=[]\n",
    "    X=[]\n",
    "    for filename in test_train_set:\n",
    "        if filename not in pssm_files:\n",
    "            continue\n",
    "        profile=open(jpred_pssm_path+filename+\".pssm\")\n",
    "        dssp=open(jpred_dssp_path+filename+\".dssp\")\n",
    "        sec_str,profile_list=reading_files(dssp,profile) #function recall\n",
    "        ss_class=class_assign(sec_str) #function recall\n",
    "        profile_list=add_window_spaces(profile_list,window) #function recall\n",
    "        for el in range(len(ss_class)):\n",
    "            window_list=profile_list[el:el+window]\n",
    "            window_list=[el for li in window_list for el in li]\n",
    "            y.append(ss_class[el])\n",
    "            X.append(window_list)\n",
    "    #print(len(y))\n",
    "    #print(len(X))\n",
    "    return(X,y)\n",
    "\n",
    "pssm_files=[]\n",
    "for pssm_file in os.listdir(jpred_pssm_path):\n",
    "    pssm_files.append(pssm_file[:-5])    \n",
    "cv_set=[]\n",
    "for file_name in os.listdir(cross_val_path):\n",
    "    cv_set.append(file_name)\n",
    "for test in cv_set:\n",
    "    test_set=[]\n",
    "    test_ids=open(cross_val_path+test)\n",
    "    for line in test_ids:\n",
    "        line=line.rstrip()\n",
    "        test_set.append(line)\n",
    "    train_set=[]\n",
    "    for train in cv_set:\n",
    "        if train!=test:\n",
    "            train_ids=open(cross_val_path+train)\n",
    "            for line in train_ids:\n",
    "                line=line.rstrip()\n",
    "                train_set.append(line) \n",
    "#print(test_set)\n",
    "#print(train_set)\n",
    "X_test,y_test=input_SVM(test_set,pssm_files,jpred_pssm_path,jpred_dssp_path)\n",
    "X_train,y_train=input_SVM(train_set,pssm_files,jpred_pssm_path,jpred_dssp_path)\n",
    "print(\"test\", len(X_test), len(y_test))\n",
    "print(\"train\", len(X_train), len(y_train))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCC:\t 0.5373618210105681 ± 0.05284331356513004\n",
      "Sen:\t 0.6530200952340268 ± 0.13578028435424747\n",
      "PVV:\t 0.7475326320565601 ± 0.06546670232946447\n"
     ]
    }
   ],
   "source": [
    "#Put here the three binary confusion matrix\n",
    "dict_mat={\n",
    "\"m_0\":[[ 48619,  22368],\n",
    " [  9555, 118749]],\n",
    "\"m_1\":[[ 17706,  26172],\n",
    " [  4786, 150627]],\n",
    "\"m_2\":[[73504, 10922],\n",
    " [45121, 69744]]\n",
    "}\n",
    "li=[\"0\",\"1\",\"2\"]\n",
    "mcc_tot=[]\n",
    "sen_tot=[]\n",
    "ppv_tot=[]\n",
    "for el in li:\n",
    "    #mat=\"m\"+\"_\"+el\n",
    "    m=dict_mat[\"m\"+\"_\"+el]    \n",
    "    mcc=((m[0][0]*m[1][1])-(m[1][0]*m[0][1]))/math.sqrt((m[0][0]+m[1][0])*(m[0][0]+m[0][1])*(m[1][1]+m[1][0])*(m[1][1]+m[0][1]))\n",
    "    ppv=(m[0][0])/(m[0][0]+m[1][0])\n",
    "    sen=(m[0][0])/sum(m[0])\n",
    "    mcc_tot.append(mcc)\n",
    "    sen_tot.append(sen)\n",
    "    ppv_tot.append(ppv)\n",
    "print(\"MCC:\\t\",statistics.mean(mcc_tot),\"±\",statistics.stdev(mcc_tot)/math.sqrt(3))\n",
    "print(\"Sen:\\t\",statistics.mean(sen_tot),\"±\",statistics.stdev(sen_tot)/math.sqrt(3))\n",
    "print(\"PVV:\\t\",statistics.mean(ppv_tot),\"±\",statistics.stdev(ppv_tot)/math.sqrt(3))     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6827434763629812 ± 0.12359005503236356\n"
     ]
    }
   ],
   "source": [
    "print(statistics.mean([0.44590943652629705,0.8624445340863252,0.7398764584763212]),\"±\",statistics.stdev([0.44590943652629705,0.8624445340863252,0.7398764584763212])/math.sqrt(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred=joblib.load(\"Pred_Blind.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20908\n"
     ]
    }
   ],
   "source": [
    "print(len(y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_path=\"Blindset_pssm/\"\n",
    "dssp_path=\"Blindset_DSSP/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20908\n",
      "20908\n",
      "CPU times: user 805 ms, sys: 0 ns, total: 805 ms\n",
      "Wall time: 831 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "window=17\n",
    "def reading_files(dssp,profile):\n",
    "    profile_list=[]\n",
    "    for line in dssp:\n",
    "        if line.startswith(\">\"):\n",
    "            sec_str1=\"\"\n",
    "            sec_str=dssp.readline().rstrip()\n",
    "            for el in sec_str: \n",
    "                if el==\"-\":\n",
    "                    sec_str1=sec_str1+\"C\"\n",
    "                else:\n",
    "                    sec_str1=sec_str1+el \n",
    "    for line in profile:\n",
    "        if \"A\" not in line:\n",
    "            line=line.rstrip()\n",
    "            line=line.split()[1:]\n",
    "            profile_list.append(line)\n",
    "    profile_list=[[float(j) for j in i] for i in profile_list] #COOL\n",
    "    return(sec_str1,profile_list)\n",
    "\n",
    "def class_assign(sec_str):\n",
    "    ss_class=[]\n",
    "    for ss in sec_str:\n",
    "        if ss==\"H\":\n",
    "            ss_class.append(1)\n",
    "        if ss==\"E\":\n",
    "            ss_class.append(2)\n",
    "        if ss==\"C\":\n",
    "            ss_class.append(3)\n",
    "    return(ss_class)\n",
    "\n",
    "def add_window_spaces(profile_list,window):\n",
    "    for i in range(window//2):\n",
    "        profile_list=[[0.0 for i in range(20)]]+profile_list\n",
    "        profile_list.append([0.0 for i in range(20)])\n",
    "    return(profile_list)\n",
    "\n",
    "#i=0\n",
    "y_true=[]\n",
    "X_pred=[]\n",
    "for filename in os.listdir(profile_path):\n",
    "    #print(filename)\n",
    "    #i+=1\n",
    "    filename=filename.split(\".\")\n",
    "    filename=\".\".join(filename[:-1])\n",
    "    profile=open(profile_path+filename+\".pssm\")\n",
    "    dssp=open(dssp_path+filename+\".dssp\")\n",
    "    sec_str,profile_list=reading_files(dssp,profile) #function recall\n",
    "    ss_class=class_assign(sec_str) #function recall\n",
    "    #print(len(ss_class))\n",
    "    #print(len(profile_list))\n",
    "    profile_list=add_window_spaces(profile_list,window) #function recall\n",
    "    for el in range(len(ss_class)):\n",
    "        window_list=profile_list[el:el+window]\n",
    "        window_list=[el for li in window_list for el in li]\n",
    "        y_true.append(ss_class[el])\n",
    "        X_pred.append(window_list)\n",
    "X_pred=np.asarray(X_pred)\n",
    "y_true=np.asarray(y_true)\n",
    "#print(i)\n",
    "print(len(y_true))\n",
    "print(len(X_pred))\n",
    "#print(type(y_true))\n",
    "#print(type(X_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20908\n",
      "20908\n"
     ]
    }
   ],
   "source": [
    "print(len(y_true))\n",
    "print(len(y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "conf_mat=confusion_matrix(y_true,y_pred,labels=[1,2,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[5403  120 2917]\n",
      " [ 217 2349 2206]\n",
      " [ 500  380 6816]]\n"
     ]
    }
   ],
   "source": [
    "print(conf_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.6967667878324086\n",
      "\n",
      "mat 2x2 for H\n",
      "[[ 5403  3037]\n",
      " [  717 11751]]\n",
      "mat 2x2 for E\n",
      "[[ 2349  2423]\n",
      " [  500 15636]]\n",
      "mat 2x2 for C\n",
      "[[6816  880]\n",
      " [5123 8089]]\n",
      "\n",
      "mcc C: 0.4851813614200574 mcc H: 0.6282802278114419 mcc E: 0.5642878706423657\n",
      "sen C: 0.8856548856548857 sen H: 0.6401658767772512 sen E: 0.49224643755238895\n",
      "ppv C: 0.5709020856018092 ppv H: 0.8828431372549019 ppv E: 0.8244998244998245\n"
     ]
    }
   ],
   "source": [
    "# conf_mat=[[6093,  151, 2553],\n",
    "#  [ 284, 2407, 2267],\n",
    "#  [ 700,  463, 7023]]\n",
    "\n",
    "def build_2_matrix(m3):\n",
    "    m2_H=np.array([[0,0],[0,0]])\n",
    "    m2_E=np.array([[0,0],[0,0]])\n",
    "    m2_C=np.array([[0,0],[0,0]])\n",
    "\n",
    "    #2 classes H\n",
    "    m2_H[0][0]=m3[0][0]\n",
    "    m2_H[0][1]=m3[0][1]+m3[0][2]\n",
    "    m2_H[1][0]=m3[1][0]+m3[2][0]\n",
    "    m2_H[1][1]=m3[1][1]+m3[1][2]+m3[2][1]+m3[2][2]\n",
    "\n",
    "    #2 classes E\n",
    "    m2_E[0][0]=m3[1][1]\n",
    "    m2_E[0][1]=m3[1][0]+m3[1][2]\n",
    "    m2_E[1][0]=m3[0][1]+m3[2][1]\n",
    "    m2_E[1][1]=m3[0][0]+m3[0][2]+m3[2][0]+m3[2][2]\n",
    "\n",
    "    #2 classes C\n",
    "    m2_C[0][0]=m3[2][2]\n",
    "    m2_C[0][1]=m3[2][0]+m3[2][1]\n",
    "    m2_C[1][0]=m3[0][2]+m3[1][2]\n",
    "    m2_C[1][1]=m3[0][0]+m3[0][1]+m3[1][0]+m3[1][1]\n",
    "\n",
    "    return m2_H,m2_E,m2_C\n",
    "\n",
    "def mcc(m):\n",
    "    d=(m[0][0]+m[1][0])*(m[0][0]+m[0][1])*(m[1][1]+m[1][0])*(m[1][1]+m[0][1])\n",
    "    return (m[0][0]*m[1][1]-m[0][1]*m[1][0])/math.sqrt(d)\n",
    "\n",
    "def get_acc(cm):\n",
    "    #print (sum(cm[0])+sum(cm[1])+sum(cm[2]))\n",
    "    return float(cm[0][0]+cm[1][1]+cm[2][2])/(sum(cm[0])+sum(cm[1])+sum(cm[2]))\n",
    "\n",
    "def sen(m):\n",
    "    return m[0][0]/(sum(m[0]))\n",
    "\n",
    "def ppv(m):\n",
    "    return m[0][0]/(m[0][0]+m[1][0])\n",
    "\n",
    "print(\"Accuracy:\",get_acc(conf_mat))\n",
    "m2_H,m2_E,m2_C=build_2_matrix(conf_mat)\n",
    "print()\n",
    "print(\"mat 2x2 for H\")\n",
    "print(m2_H)\n",
    "print(\"mat 2x2 for E\")\n",
    "print(m2_E)\n",
    "print(\"mat 2x2 for C\")\n",
    "print(m2_C)\n",
    "print()\n",
    "print(\"mcc C:\",mcc(m2_C),\"mcc H:\",mcc(m2_H),\"mcc E:\",mcc(m2_E))\n",
    "print(\"sen C:\",sen(m2_C),\"sen H:\",sen(m2_H),\"sen E:\",sen(m2_E))\n",
    "print(\"ppv C:\",ppv(m2_C),\"ppv H:\",ppv(m2_H),\"ppv E:\",ppv(m2_E))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCC:\t 0.6406666666666667 ± 0.09043659534600902\n"
     ]
    }
   ],
   "source": [
    "mcc_tot=[0.633,0.488,0.801]\n",
    "print(\"MCC:\\t\",statistics.mean(mcc_tot),\"±\",statistics.stdev(mcc_tot)/math.sqrt(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
